// @storybook/mcp@0.8.0/dist/index.js
import { t as instructions_default } from "./instructions-CiOmp0JI.js";
import { McpServer } from "tmcp";
import { ValibotJsonSchemaAdapter } from "@tmcp/adapter-valibot";
import { HttpTransport } from "@tmcp/transport-http";
import * as v from "valibot";

//#region package.json
var name = "@storybook/mcp";
var version = "0.8.0";
var description = "MCP server that serves knowledge about your components based on your Storybook stories and documentation";

//#endregion
//#region src/types.ts
const JSDocTag = v.record(v.string(), v.array(v.string()));
const ManifestError = v.object({
	name: v.string(),
	message: v.string()
});
const BaseManifest = v.object({
	name: v.string(),
	description: v.optional(v.string()),
	jsDocTags: v.optional(JSDocTag),
	error: v.optional(ManifestError)
});
const Story = v.object({
	...BaseManifest.entries,
	id: v.optional(v.string()),
	snippet: v.optional(v.string()),
	summary: v.optional(v.string())
});
/**
* A JSON Reference (`{ $ref }`) pointing at a value in another manifest document.
* Used by the v1 (split/ref) manifest format for docgen, story-docs and MDX payloads.
*/
const JsonRef = v.object({ $ref: v.string() });
/** Inline (v0) docs entry: the full MDX `content` is embedded. */
const DocV0 = v.object({
	id: v.string(),
	name: v.string(),
	title: v.optional(v.string()),
	path: v.optional(v.string()),
	content: v.optional(v.string()),
	summary: v.optional(v.string()),
	error: v.optional(ManifestError)
});
const BaseInlineComponentProperties = v.object({
	...BaseManifest.entries,
	path: v.optional(v.string()),
	summary: v.optional(v.string()),
	import: v.optional(v.string()),
	reactDocgen: v.optional(v.any()),
	reactDocgenTypescript: v.optional(v.any()),
	reactComponentMeta: v.optional(v.any())
});
const SubcomponentManifest = v.object({ ...BaseInlineComponentProperties.entries });
/** Inline (v0) component: docgen, stories and attached docs are all embedded. */
const ComponentManifestV0 = v.object({
	...BaseInlineComponentProperties.entries,
	id: v.string(),
	stories: v.optional(v.array(Story)),
	subcomponents: v.optional(v.record(v.string(), SubcomponentManifest)),
	docs: v.optional(v.record(v.string(), DocV0))
});
const ComponentManifestMapV0 = v.object({
	v: v.literal(0),
	components: v.record(v.string(), ComponentManifestV0)
});
const DocsManifestMapV0 = v.object({
	v: v.literal(0),
	docs: v.record(v.string(), DocV0)
});
/**
* Shallow (v1) docs entry. The full MDX payload (title/path/content) lives behind
* `mdx.$ref`; `mdx` is optional because Storybook's in-process dev index omits it
* (those entries are resolved in-process via {@link StorybookContext.resolveEntry}).
*/
const DocV1 = v.object({
	id: v.string(),
	name: v.string(),
	summary: v.optional(v.string()),
	mdx: v.optional(JsonRef),
	error: v.optional(ManifestError)
});
/**
* Shallow (v1) component index row. Identity + summary are inlined for cheap
* listing; docgen and story-docs live behind `$ref`s, attached docs behind nested
* `mdx.$ref`s. `docgen`/`stories` are optional (the in-process dev index omits
* `docgen`, and docs-only components have neither).
*/
const ComponentManifestV1 = v.object({
	id: v.string(),
	name: v.string(),
	description: v.optional(v.string()),
	summary: v.optional(v.string()),
	error: v.optional(ManifestError),
	docgen: v.optional(JsonRef),
	stories: v.optional(JsonRef),
	docs: v.optional(v.record(v.string(), DocV1))
});
const ComponentManifestMapV1 = v.object({
	v: v.literal(1),
	components: v.record(v.string(), ComponentManifestV1)
});
const DocsManifestMapV1 = v.object({
	v: v.literal(1),
	docs: v.record(v.string(), DocV1)
});
/** `components.json`, discriminated on `v` (0 = inline, 1 = split/ref). */
const ComponentManifestMap = v.variant("v", [ComponentManifestMapV0, ComponentManifestMapV1]);
/**
* `docs.json` for unattached/standalone documentation entries (served at
* `/manifests/docs.json`), discriminated on `v`.
*/
const DocsManifestMap = v.variant("v", [DocsManifestMapV0, DocsManifestMapV1]);
/**
* Shared Valibot field for the storybookId input, used in multi-source mode.
* Reused across tools that support source selection.
*/
const StorybookIdField = { storybookId: v.pipe(v.string(), v.description("The Storybook source ID (e.g., \"local\", \"tetra\"). Required when multiple Storybooks are composed. See list-all-documentation for available sources.")) };

//#endregion
//#region src/utils/adapt-core-manifest.ts
/**
* Adapts Storybook's `experimentalDocgenServer` open-service payloads (the "core
* format") into `@storybook/mcp`'s internal {@link ComponentManifest}/{@link Doc}
* shapes. This is the single place that bridges the two formats, used by both the
* ref-resolution path (built/static/remote manifests) and the addon's in-process
* `resolveEntry` hook (dev).
*
* Only `argTypes` is dropped: prop types still come from `reactComponentMeta` and
* the `react*` docgen-engine fields, which are passed through unchanged.
*/
const ARG_TYPES_KEY = "argTypes";
/** Converts the story-docs `stories` record (or an already-resolved array) into `Story[]`. */
function adaptCoreStories(stories) {
	if (!stories) return;
	if (Array.isArray(stories)) return stories;
	return Object.values(stories);
}
/** Adapts one MDX service payload into a {@link Doc}. */
function adaptCoreDoc(doc) {
	return { ...doc };
}
/** Adapts a core-format component (docgen + story-docs + attached MDX) into a {@link ComponentManifest}. */
function adaptCoreComponent(core) {
	const { stories, docs, [ARG_TYPES_KEY]: _argTypes, ...rest } = core;
	const component = { ...rest };
	const adaptedStories = adaptCoreStories(stories);
	if (adaptedStories) component.stories = adaptedStories;
	if (docs) component.docs = Object.fromEntries(Object.entries(docs).map(([id, doc]) => [id, adaptCoreDoc(doc)]));
	return component;
}

//#endregion
//#region src/utils/requires-own-mcp.ts
function getSourceMcpEndpoint(source) {
	const base = new URL(source.url);
	base.pathname = `${base.pathname.replace(/\/$/, "")}/`;
	return new URL("mcp", base).toString();
}
function formatRequiresOwnMcpNotice(source, endpoint, options = {}) {
	const parts = [];
	if (options.includeHeader ?? true) {
		parts.push(`# ${source.title}`);
		parts.push(`id: ${source.id}`);
		parts.push("");
	}
	parts.push("This composed Storybook is private and cannot be read through the local Storybook MCP proxy.");
	parts.push("");
	parts.push("Use this source's own MCP endpoint instead:");
	parts.push(endpoint);
	return parts.join("\n");
}

//#endregion
//#region src/utils/get-manifest.ts
/**
* The paths to the manifest files relative to the Storybook build
*/
const COMPONENT_MANIFEST_PATH = "./manifests/components.json";
const DOCS_MANIFEST_PATH = "./manifests/docs.json";
/** Empty placeholder manifest used for sources that errored or need their own MCP endpoint. */
const EMPTY_COMPONENT_MANIFEST = {
	v: 1,
	components: {}
};
/**
* Error thrown when getting or parsing a manifest fails
*/
var ManifestGetError = class extends Error {
	url;
	cause;
	constructor(message, url, cause) {
		super(message);
		this.name = "ManifestGetError";
		this.url = url ?? "No source URL provided";
		this.cause = cause;
	}
};
var RequiresOwnMcpError = class extends Error {
	source;
	endpoint;
	constructor(source) {
		const endpoint = getSourceMcpEndpoint(source);
		super(`Composed Storybook "${source.title}" requires its own MCP endpoint: ${endpoint}`);
		this.name = "RequiresOwnMcpError";
		this.source = source;
		this.endpoint = endpoint;
	}
};
/**
* Converts an error to MCP-compatible content format
*
* @param error - The error to convert (can be any type)
* @returns A tool result with error content and isError flag
*/
const errorToMCPContent = (error) => {
	if (error instanceof RequiresOwnMcpError) return { content: [{
		type: "text",
		text: formatRequiresOwnMcpNotice(error.source, error.endpoint)
	}] };
	let fullMessage = `${error instanceof ManifestGetError ? "Error getting manifest" : "Unexpected error"}: ${error instanceof Error ? error.message : String(error)}`;
	if (error instanceof ManifestGetError && error.cause) {
		const causeMessage = error.cause instanceof Error ? error.cause.message : String(error.cause);
		fullMessage += `\nCaused by: ${causeMessage}`;
	}
	return {
		content: [{
			type: "text",
			text: fullMessage
		}],
		isError: true
	};
};
/**
* Parses a JSON string and validates it against a Valibot schema
*/
function parseManifest({ jsonString, schema, name, url }) {
	try {
		return v.parse(v.pipe(v.string(), v.parseJson(), schema), jsonString);
	} catch (error) {
		throw new ManifestGetError(`Failed to parse ${name} manifest:
${error instanceof v.ValiError ? error.issues.map((i) => i.message).join("\n") : String(error)}`, url);
	}
}
/**
* Gets component and docs manifest from a request or using a custom provider
*
* @param request - The HTTP request to get the manifest for (optional when using custom manifestProvider)
* @param manifestProvider - Optional custom function to get the manifest
* @param source - Optional source for multi-source mode
* @returns A promise that resolves to the parsed ComponentManifestMap
* @throws {ManifestGetError} If getting the manifest fails or the response is invalid
*/
async function getManifests(request, manifestProvider, source) {
	const provider = manifestProvider ?? defaultManifestProvider;
	const [componentResult, docsResult] = await Promise.allSettled([provider(request, COMPONENT_MANIFEST_PATH, source), provider(request, DOCS_MANIFEST_PATH, source)]);
	const getUrl = (path) => request ? getManifestUrlFromRequest(request, path) : "Unknown manifest source";
	if (componentResult.status === "rejected") {
		const reason = componentResult.reason;
		if (reason instanceof RequiresOwnMcpError) throw reason;
		const hint = reason instanceof ManifestGetError && reason.message.includes("404") ? `\nHint: The Storybook at this URL may not have the component manifest enabled. Add \`features: { componentsManifest: true }\` (or \`features: { experimentalComponentsManifest: true }\` for older Storybook versions) to its main.ts config.` : "";
		throw new ManifestGetError(`Failed to get component manifest: ${reason instanceof Error ? reason.message : String(reason)}${hint}`, getUrl(COMPONENT_MANIFEST_PATH), reason instanceof Error ? reason : void 0);
	}
	const componentManifest = parseManifest({
		jsonString: componentResult.value,
		schema: ComponentManifestMap,
		name: "component",
		url: getUrl(COMPONENT_MANIFEST_PATH)
	});
	if (Object.keys(componentManifest.components).length === 0) throw new ManifestGetError(`No components found in the manifest`, getUrl(COMPONENT_MANIFEST_PATH));
	if (docsResult.status === "rejected") return { componentManifest };
	return {
		componentManifest,
		docsManifest: parseManifest({
			jsonString: docsResult.value,
			schema: DocsManifestMap,
			name: "docs",
			url: getUrl(DOCS_MANIFEST_PATH)
		})
	};
}
/**
* Resolves a `$ref` into the provider path of the referenced file and the
* JSON-pointer segments into it.
*
* The `$ref` file path is relative to the component manifest's location, e.g.
* `"../services/core/docgen/button.json#/components/button"` from
* `./manifests/components.json` resolves to `./services/core/docgen/button.json`
* with pointer `["components", "button"]`. The same base (`manifests/`) applies to
* docgen, story-docs and MDX refs.
*/
function parseManifestRef(ref) {
	const [filePath = "", hash = ""] = ref.split("#");
	const manifestDir = COMPONENT_MANIFEST_PATH.replace(/^\.\//, "").replace(/[^/]+$/, "");
	const resolved = new URL(filePath, `https://localhost/${manifestDir}`).pathname.replace(/^\//, "");
	const pointer = hash.split("/").filter(Boolean).map((segment) => segment.replace(/~1/g, "/").replace(/~0/g, "~"));
	return {
		path: `./${resolved}`,
		pointer
	};
}
/**
* Fetches the file a `$ref` points at (verbatim — never a fabricated path) and
* walks its JSON pointer, returning the pointed-to value.
*/
async function fetchRefValue(ref, request, provider, source) {
	const { path, pointer } = parseManifestRef(ref);
	const jsonString = await provider(request, path, source);
	let target;
	try {
		target = JSON.parse(jsonString);
	} catch (error) {
		throw new ManifestGetError(`Failed to parse externalized payload referenced by "${ref}"`, path, error instanceof Error ? error : void 0);
	}
	for (const key of pointer) if (target && typeof target === "object" && key in target) target = target[key];
	else throw new ManifestGetError(`Reference "${ref}" could not be resolved: missing "${key}".`, path);
	return target;
}
/**
* Resolves a component index entry into a full component manifest by following any
* `$ref`s it carries (docgen, story-docs, attached MDX docs), then adapting the
* core-format payloads into `@storybook/mcp`'s internal shape.
*
* Refs are read verbatim from the entry — never fabricated from an id — so a
* provider is only ever asked for top-level manifest paths or paths a manifest
* already pointed at. Components without any `$ref` (inline/v0 format, or entries
* already resolved in-process) are returned unchanged.
*/
async function resolveComponentEntry(component, request, manifestProvider, source) {
	const docgenRef = "docgen" in component ? component.docgen?.$ref : void 0;
	const storiesRef = component.stories && !Array.isArray(component.stories) ? component.stories.$ref : void 0;
	const docEntries = component.docs ? Object.entries(component.docs) : [];
	const docRefs = docEntries.filter(([, doc]) => "mdx" in doc && !!doc.mdx?.$ref);
	if (!docgenRef && !storiesRef && docRefs.length === 0) return component;
	const provider = manifestProvider ?? defaultManifestProvider;
	let core = {
		id: component.id,
		name: component.name,
		...component.description !== void 0 ? { description: component.description } : {},
		...component.summary !== void 0 ? { summary: component.summary } : {},
		...component.error !== void 0 ? { error: component.error } : {}
	};
	if (docgenRef) {
		const payload = await fetchRefValue(docgenRef, request, provider, source);
		core = {
			...core,
			...payload,
			id: component.id,
			name: component.name,
			...component.description !== void 0 ? { description: component.description } : {},
			...component.summary !== void 0 ? { summary: component.summary } : {},
			...component.error !== void 0 ? { error: component.error } : {}
		};
	}
	if (Array.isArray(component.stories)) core.stories = component.stories;
	if (storiesRef) {
		const storyDocs = await fetchRefValue(storiesRef, request, provider, source);
		if (storyDocs?.stories) core.stories = storyDocs.stories;
		if (storyDocs?.import) core.import = storyDocs.import;
	}
	if (docEntries.length > 0) {
		const docs = {};
		for (const [docId, doc] of docEntries) {
			const mdxRef = "mdx" in doc ? doc.mdx?.$ref : void 0;
			docs[docId] = mdxRef ? await fetchRefValue(mdxRef, request, provider, source) : doc;
		}
		core.docs = docs;
	}
	return adaptCoreComponent(core);
}
/**
* Resolves only a component's stories `$ref` (story-docs payload) into a `Story[]`,
* leaving docgen/docs refs untouched. Used by `list-all-documentation` when story
* ids are requested, so listing doesn't pay for docgen/MDX resolution it won't show.
*/
async function resolveComponentStories(component, request, manifestProvider, source) {
	if (!component.stories || Array.isArray(component.stories)) return component;
	const provider = manifestProvider ?? defaultManifestProvider;
	const storyDocs = await fetchRefValue(component.stories.$ref, request, provider, source);
	return {
		...component,
		stories: storyDocs?.stories ? adaptCoreStories(storyDocs.stories) : []
	};
}
/**
* Resolves a standalone docs entry, following its `mdx.$ref` when present.
* Inline docs (v0) are returned unchanged.
*/
async function resolveDoc(doc, request, manifestProvider, source) {
	const ref = "mdx" in doc ? doc.mdx?.$ref : void 0;
	if (!ref) return doc;
	const payload = await fetchRefValue(ref, request, manifestProvider ?? defaultManifestProvider, source);
	return adaptCoreDoc({
		...payload,
		id: payload.id ?? doc.id,
		name: payload.name ?? doc.name
	});
}
/**
* Constructs the manifest URL from the request origin and the top-level manifest path.
*/
function getManifestUrlFromRequest(request, path) {
	const normalizedPath = path.replace(/^\.\//, "");
	return new URL(`/${normalizedPath}`, request.url).toString();
}
/**
* Default manifest provider that fetches from the same origin as the request,
* using Storybook's top-level manifest path.
*/
async function defaultManifestProvider(request, path) {
	if (!request) throw new ManifestGetError("Request is required when using the default manifest provider. You must either pass the original request forward to the server context, or set a custom manifestProvider that doesn't need the request.");
	const manifestUrl = getManifestUrlFromRequest(request, path);
	const response = await fetch(manifestUrl);
	if (!response.ok) throw new ManifestGetError(`Failed to fetch manifest: ${response.status} ${response.statusText}`, manifestUrl);
	const contentType = response.headers.get("content-type");
	if (!contentType?.includes("application/json")) throw new ManifestGetError(`Invalid content type: expected application/json, got ${contentType}`, manifestUrl);
	return response.text();
}
/**
* Gets manifests from multiple sources.
* Returns an array of source manifests, each containing the source info and its manifests.
* Failures for individual sources are captured as errors rather than failing the entire request.
*
* @param sources - Array of source configurations
* @param request - The HTTP request (used for local source)
* @param manifestProvider - Function to fetch manifests, receives source as third parameter
* @returns Promise resolving to array of source manifests
* @throws {ManifestGetError} If no sources could be fetched successfully
*/
async function getMultiSourceManifests(sources, request, manifestProvider) {
	const results = await Promise.all(sources.map(async (source) => {
		try {
			const manifests = await getManifests(request, manifestProvider, source);
			return {
				source,
				componentManifest: manifests.componentManifest,
				docsManifest: manifests.docsManifest
			};
		} catch (error) {
			if (error instanceof RequiresOwnMcpError) return {
				source,
				componentManifest: EMPTY_COMPONENT_MANIFEST,
				notice: {
					kind: "requires-own-mcp",
					endpoint: error.endpoint
				}
			};
			return {
				source,
				componentManifest: EMPTY_COMPONENT_MANIFEST,
				error: error instanceof Error ? error.message : String(error)
			};
		}
	}));
	const successCount = results.filter((r) => !r.error && !r.notice).length;
	const noticeCount = results.filter((r) => r.notice).length;
	if (successCount === 0 && noticeCount === 0) throw new ManifestGetError(`Failed to fetch manifests from any source. Errors:\n${results.map((r) => `- ${r.source.title}: ${r.error}`).join("\n")}`);
	return results;
}

//#endregion
//#region src/utils/map-with-concurrency.ts
/**
* Maps `items` with at most `limit` concurrent in-flight `fn` calls.
*/
async function mapWithConcurrency(items, limit, fn) {
	if (items.length === 0) return [];
	const safeLimit = Number.isFinite(limit) ? limit : items.length;
	const concurrency = Math.max(1, Math.min(safeLimit, items.length));
	const results = Array(items.length);
	let nextIndex = 0;
	async function worker() {
		while (nextIndex < items.length) {
			const index = nextIndex++;
			results[index] = await fn(items[index], index);
		}
	}
	await Promise.all(Array.from({ length: concurrency }, () => worker()));
	return results;
}

//#endregion
//#region src/utils/parse-react-docgen.ts
function serializeTsType(tsType) {
	if (!tsType) return void 0;
	if ("raw" in tsType && typeof tsType.raw === "string" && tsType.raw.trim().length > 0) return tsType.raw;
	if (!tsType.name) return void 0;
	if ("elements" in tsType) {
		const serializeElements = () => (tsType.elements ?? []).map((el) => serializeTsType(el) ?? "unknown");
		switch (tsType.name) {
			case "union": return serializeElements().join(" | ");
			case "intersection": return serializeElements().join(" & ");
			case "Array": return `${serializeTsType((tsType.elements ?? [])[0]) ?? "unknown"}[]`;
			case "tuple": return `[${serializeElements().join(", ")}]`;
		}
	}
	if ("value" in tsType && tsType.name === "literal") return tsType.value;
	if ("signature" in tsType && tsType.name === "signature") {
		if (tsType.type === "function") {
			const args = (tsType.signature?.arguments ?? []).map((a) => {
				const argType = serializeTsType(a.type) ?? "any";
				return `${a.name}: ${argType}`;
			});
			const ret = serializeTsType(tsType.signature?.return) ?? "void";
			return `(${args.join(", ")}) => ${ret}`;
		}
		if (tsType.type === "object") return `{ ${(tsType.signature?.properties ?? []).map((p) => {
			const req = Boolean(p.value?.required);
			const propType = serializeTsType(p.value) ?? "any";
			return `${p.key}${req ? "" : "?"}: ${propType}`;
		}).join("; ")} }`;
		return "unknown";
	}
	if ("elements" in tsType) {
		const inner = (tsType.elements ?? []).map((el) => serializeTsType(el) ?? "unknown");
		if (inner.length > 0) return `${tsType.name}<${inner.join(", ")}>`;
	}
	return tsType.name;
}
const parseReactDocgen = (reactDocgen) => {
	const props = reactDocgen?.props ?? {};
	return { props: Object.fromEntries(Object.entries(props).map(([propName, prop]) => [propName, {
		description: prop.description,
		type: serializeTsType(prop.tsType ?? prop.type),
		defaultValue: prop.defaultValue?.value,
		required: prop.required
	}])) };
};
/**
* Parse react-docgen-typescript output into the same simplified ParsedDocgen format.
* RDT uses flat type strings (prop.type.name / prop.type.raw) instead of react-docgen's
* nested tsType structure, so no serialization is needed.
*/
const parseComponentDocLike = (componentDoc) => {
	const props = componentDoc.props ?? {};
	return { props: Object.fromEntries(Object.entries(props).map(([propName, prop]) => [propName, {
		description: prop.description || void 0,
		type: prop.type?.raw ?? prop.type?.name,
		defaultValue: prop.defaultValue?.value,
		required: prop.required
	}])) };
};
const parseReactDocgenTypescript = (reactDocgenTypescript) => parseComponentDocLike(reactDocgenTypescript);
const parseReactComponentMeta = (reactComponentMeta) => parseComponentDocLike(reactComponentMeta);

//#endregion
//#region src/utils/dedent.ts
/**
* Copied from https://github.com/tamino-martinius/node-ts-dedent/blob/18c4736c79806d7fe78bdaaaf5ae307b79f9574a/src/index.ts
*
* Rationale for vendoring instead of using the ts-dedent package:
* 1. The package has a broken ESM distribution - the IDE auto-imports as
*    `import dedent from 'ts-dedent'` but that breaks at runtime, requiring
*    manual changes to named imports. See:
*    - https://github.com/tamino-martinius/node-ts-dedent/issues/42
*    - https://github.com/tamino-martinius/node-ts-dedent/pull/41
* 2. Keeps bundle size smaller for such a simple utility (avoid "is-even" syndrome)
*/
function dedent(templ, ...values) {
	let strings = Array.from(typeof templ === "string" ? [templ] : templ);
	strings[strings.length - 1] = strings.at(-1).replace(/\r?\n([\t ]*)$/, "");
	const indentLengths = strings.reduce((arr, str) => {
		const matches = str.match(/\n([\t ]+|(?!\s).)/g);
		if (matches) return arr.concat(matches.map((match) => match.match(/[\t ]/g)?.length ?? 0));
		return arr;
	}, []);
	if (indentLengths.length) {
		const pattern = new RegExp(`\n[\t ]{${Math.min(...indentLengths)}}`, "g");
		strings = strings.map((str) => str.replace(pattern, "\n"));
	}
	strings[0] = strings[0].replace(/^\r?\n/, "");
	let string = strings[0];
	values.forEach((value, i) => {
		const endentations = string.match(/(?:^|\n)( *)$/);
		const endentation = endentations ? endentations[1] : "";
		let indentedValue = value;
		if (typeof value === "string" && value.includes("\n")) indentedValue = String(value).split("\n").map((str, i) => {
			return i === 0 ? str : `${endentation}${str}`;
		}).join("\n");
		string += `${String(indentedValue)}${strings[i + 1]}`;
	});
	return string;
}

//#endregion
//#region src/utils/manifest-formatter/extract-docs-summary.ts
/**
* Maximum length for a summary before truncation.
*/
const MAX_SUMMARY_LENGTH = 90;
/**
* Extracts a summary from MDX content.
* The summary is created by:
* 1. Removing import statements
* 2. Removing JSX/MDX expressions
* 3. Extracting only text content from JSX/HTML elements
* 4. Truncating to MAX_SUMMARY_LENGTH characters if needed
*
* @param content - The MDX content string
* @returns A summary string, or undefined if no meaningful text content is found
*/
function extractDocsSummary(content) {
	let result = content;
	result = result.replace(/^\s*import\s+(?:[\s\S]*?from\s+)?['"][^'"]+['"];?\s*$/gm, "");
	let prevResult = "";
	while (prevResult !== result) {
		prevResult = result;
		result = result.replace(/\{[^{}]*\}/g, "");
	}
	result = result.replace(/<[^>]+\/>/g, "");
	prevResult = "";
	while (prevResult !== result) {
		prevResult = result;
		result = result.replace(/<(\w+)[^>]*>([\s\S]*?)<\/\1>/g, "$2");
	}
	result = result.replace(/<[^>]+>/g, "");
	result = result.replace(/\s+/g, " ").trim();
	if (!result) return;
	if (result.length > MAX_SUMMARY_LENGTH) return `${result.slice(0, MAX_SUMMARY_LENGTH)}...`;
	return result;
}

//#endregion
//#region src/utils/manifest-formatter/markdown.ts
/**
* Maximum number of stories to show in full detail in component manifests.
* Remaining stories will be shown as names only.
*/
const MAX_STORIES_TO_SHOW = 3;
function formatComponentLine(component) {
	const summary = component.summary ?? (component.description ? component.description.length > MAX_SUMMARY_LENGTH ? `${component.description.slice(0, MAX_SUMMARY_LENGTH)}...` : component.description : void 0);
	if (summary) return `- ${component.name} (${component.id}): ${summary}`;
	return `- ${component.name} (${component.id})`;
}
function formatDocLine(doc) {
	const title = "title" in doc ? doc.title : void 0;
	const content = "content" in doc ? doc.content : void 0;
	const summary = doc.summary ?? extractDocsSummary(content ?? "");
	return `- ${title ?? doc.name} (${doc.id})${summary ? `: ${summary}` : ""}`;
}
function formatStorySubLine(story) {
	return `  - ${story.name}` + (story.id ? ` (${story.id})` : "");
}
/**
* Extracts a summary from an object with optional summary and description fields.
* Prefers summary if available, otherwise truncates description to maxLength.
*/
function extractSummary(item, maxLength = MAX_SUMMARY_LENGTH) {
	if (item.summary) return item.summary;
	if (item.description) return item.description.length > maxLength ? `${item.description.slice(0, maxLength)}...` : item.description;
}
/**
* Extract parsed docgen from a component manifest, preferring reactDocgen over
* reactDocgenTypescript over reactComponentMeta.
*/
function getParsedDocgen(componentManifest) {
	if (componentManifest.reactDocgen) return parseReactDocgen(componentManifest.reactDocgen);
	if (componentManifest.reactDocgenTypescript) return parseReactDocgenTypescript(componentManifest.reactDocgenTypescript);
	if (componentManifest.reactComponentMeta) return parseReactComponentMeta(componentManifest.reactComponentMeta);
}
/**
* Formats a story's content (description + code snippet) into markdown.
* Reusable helper for both formatComponentManifest and formatStoryDocumentation.
*/
function formatStoryContent(story, importStatement) {
	const parts = [];
	if (story.description) {
		parts.push(story.description);
		parts.push("");
	}
	parts.push("```");
	if (importStatement) {
		parts.push(importStatement);
		parts.push("");
	}
	parts.push(story.snippet ?? "");
	parts.push("```");
	return parts;
}
function formatPropsSection(parsedDocgen, options = {}) {
	const propEntries = parsedDocgen ? Object.entries(parsedDocgen.props) : [];
	if (propEntries.length === 0) return [];
	const title = options.title ?? "## Props";
	const typeName = options.typeName ?? "Props";
	const parts = [];
	parts.push(title);
	parts.push("");
	parts.push("```");
	parts.push(`export type ${typeName} = {`);
	for (const [propName, propInfo] of propEntries) {
		const type = propInfo.type ?? "any";
		const isRequired = propInfo.required ?? true;
		const hasDefault = propInfo.defaultValue !== void 0;
		if (propInfo.description !== void 0) {
			parts.push("  /**");
			parts.push(`    ${propInfo.description}`);
			parts.push("  */");
		}
		let propLine = `  ${propName}`;
		if (!isRequired) propLine += "?";
		propLine += `: ${type}`;
		if (hasDefault) propLine += ` = ${propInfo.defaultValue}`;
		propLine += ";";
		parts.push(propLine);
	}
	parts.push("}");
	parts.push("```");
	parts.push("");
	return parts;
}
function formatSubcomponentsSection(subcomponents) {
	if (!subcomponents || Object.keys(subcomponents).length === 0) return [];
	const parts = [];
	parts.push("## Subcomponents");
	parts.push("");
	for (const [key, subcomponent] of Object.entries(subcomponents)) {
		parts.push(`### ${subcomponent.name || key}`);
		parts.push("");
		if (subcomponent.summary) {
			parts.push(subcomponent.summary);
			parts.push("");
		}
		if (subcomponent.description) {
			parts.push(subcomponent.description);
			parts.push("");
		}
		if (subcomponent.import) {
			parts.push("```");
			parts.push(subcomponent.import);
			parts.push("```");
			parts.push("");
		}
		if (subcomponent.error) {
			parts.push(`Error: ${subcomponent.error.name}`);
			parts.push("");
			parts.push("```");
			parts.push(subcomponent.error.message);
			parts.push("```");
			parts.push("");
			continue;
		}
		const parsedDocgen = getParsedDocgen(subcomponent);
		const typeName = `${(subcomponent.name || key).replace(/\W+/g, "")}Props`;
		parts.push(...formatPropsSection(parsedDocgen, {
			title: "#### Props",
			typeName
		}));
	}
	return parts;
}
/**
* Format a single component manifest into markdown.
*/
function formatComponentManifest(componentManifest) {
	const parts = [];
	parts.push(`# ${componentManifest.name}`);
	parts.push("");
	parts.push(`ID: ${componentManifest.id}`);
	parts.push("");
	if (componentManifest.description) {
		parts.push(componentManifest.description);
		parts.push("");
	}
	parts.push(...formatSubcomponentsSection(componentManifest.subcomponents));
	const parsedDocgen = getParsedDocgen(componentManifest);
	const stories = Array.isArray(componentManifest.stories) ? componentManifest.stories : [];
	if (stories.length > 0) {
		parts.push("## Stories");
		parts.push("");
		const storiesWithSnippets = stories.filter((s) => s.snippet);
		const hasProps = parsedDocgen && Object.keys(parsedDocgen.props).length > 0;
		const storiesToShow = hasProps ? storiesWithSnippets.slice(0, MAX_STORIES_TO_SHOW) : storiesWithSnippets;
		const remainingStories = hasProps ? storiesWithSnippets.slice(MAX_STORIES_TO_SHOW) : [];
		for (const story of storiesToShow) {
			parts.push(`### ${story.name}`);
			parts.push("");
			if (story.id) {
				parts.push(`Story ID: ${story.id}`);
				parts.push("");
			}
			parts.push(...formatStoryContent(story, componentManifest.import));
			parts.push("");
		}
		if (remainingStories.length > 0) {
			if (storiesToShow.length > 0) parts.push("### Other Stories");
			parts.push("");
			for (const story of remainingStories) {
				const summary = extractSummary(story);
				const summaryPart = summary ? `: ${summary}` : "";
				const storyLabel = story.id ? `${story.name} (${story.id})` : story.name;
				parts.push(`- ${storyLabel}${summaryPart}`);
			}
			parts.push("");
		}
	}
	parts.push(...formatPropsSection(parsedDocgen));
	if (componentManifest.docs && Object.keys(componentManifest.docs).length > 0) {
		const docsWithContent = Object.values(componentManifest.docs).filter((doc) => (doc.content ?? "").trim().length > 0);
		if (docsWithContent.length > 0) {
			parts.push("## Docs");
			parts.push("");
			for (const doc of docsWithContent) {
				parts.push(`### ${doc.name}`);
				parts.push("");
				parts.push(doc.content ?? "");
				parts.push("");
			}
		}
	}
	return parts.join("\n").trim();
}
/**
* Format a single doc manifest into markdown.
*/
function formatDocsManifest(doc) {
	return dedent`# ${doc.title ?? doc.name}

			${doc.content ?? ""}`;
}
/**
* Format a component manifest map into a markdown list.
* @param manifest - The component manifest map to format
* @returns Formatted string representation of the component list
*/
function formatManifestsToLists(manifests, options = {}) {
	const parts = [];
	parts.push("# Components");
	parts.push("");
	for (const component of Object.values(manifests.componentManifest.components)) {
		parts.push(formatComponentLine(component));
		if (options.withStoryIds && Array.isArray(component.stories)) for (const story of component.stories) parts.push(formatStorySubLine(story));
	}
	parts.push("");
	if (!manifests.docsManifest) return parts.join("\n").trim();
	parts.push("# Docs");
	parts.push("");
	for (const doc of Object.values(manifests.docsManifest.docs)) parts.push(formatDocLine(doc));
	return parts.join("\n").trim();
}
function formatMultiSourceManifestsToLists(manifests, options = {}) {
	const parts = [];
	for (const { source, componentManifest, docsManifest, error, notice } of manifests) {
		parts.push(`# ${source.title}`);
		parts.push(`id: ${source.id}`);
		parts.push("");
		if (error) {
			parts.push(`error: ${error}`);
			parts.push("");
			continue;
		}
		if (notice) {
			parts.push(formatRequiresOwnMcpNotice(source, notice.endpoint, { includeHeader: false }));
			parts.push("");
			continue;
		}
		const components = Object.values(componentManifest.components);
		if (components.length > 0) {
			parts.push("## Components");
			parts.push("");
			for (const component of components) {
				parts.push(formatComponentLine(component));
				if (options.withStoryIds && Array.isArray(component.stories)) for (const story of component.stories) parts.push(formatStorySubLine(story));
			}
			parts.push("");
		}
		if (docsManifest && Object.keys(docsManifest.docs).length > 0) {
			parts.push("## Docs");
			parts.push("");
			for (const doc of Object.values(docsManifest.docs)) parts.push(formatDocLine(doc));
			parts.push("");
		}
	}
	return parts.join("\n").trim();
}
/**
* Format a single story's documentation.
*/
function formatStoryDocumentation(componentManifest, storyName) {
	const story = Array.isArray(componentManifest.stories) ? componentManifest.stories.find((s) => s.name === storyName) : void 0;
	if (!story || !story.snippet) return "";
	const parts = [];
	parts.push(`# ${componentManifest.name} - ${story.name}`);
	parts.push("");
	parts.push(...formatStoryContent(story, componentManifest.import));
	return parts.join("\n").trim();
}

//#endregion
//#region src/tools/list-all-documentation.ts
const LIST_TOOL_NAME = "list-all-documentation";
const ListAllDocumentationInput = v.object({ withStoryIds: v.optional(v.pipe(v.boolean(), v.description("When true, includes story sub-bullets under each component with story name and story ID. Use this to discover IDs for downstream story-focused workflows without filesystem lookup.")), false) });
function getListAllDocumentationToolMetadata() {
	return {
		name: LIST_TOOL_NAME,
		title: "List All Documentation",
		description: "List all available UI components and documentation entries from the Storybook, returning the IDs the other documentation tools take as input. Call this first for any UI task — before writing a new component, check what the design system already provides and build on it instead of hand-rolling a duplicate; before answering any question about props, API, or usage, discover the relevant IDs here rather than reading component source. Then fetch the entries with get-documentation, referencing only IDs returned here — never guess IDs. When multiple Storybook sources are configured, entries from every source are included; scope follow-up calls to one source via their `storybookId` input. Pass `withStoryIds: true` when you need story IDs for other tools.",
		schema: ListAllDocumentationInput
	};
}
async function addListAllDocumentationTool(server, enabled) {
	server.tool({
		...getListAllDocumentationToolMetadata(),
		enabled
	}, async (input) => {
		try {
			const ctx = server.ctx.custom;
			const withStoryIds = input.withStoryIds ?? false;
			if (ctx?.sources?.some((s) => s.url)) {
				const multiSourceManifests = await getMultiSourceManifests(ctx.sources, ctx.request, ctx.manifestProvider);
				if (withStoryIds) await Promise.all(multiSourceManifests.map(async (sourceManifest) => {
					if (sourceManifest.error || sourceManifest.notice) return;
					try {
						const components = sourceManifest.componentManifest.components;
						const resolved = await mapWithConcurrency(Object.entries(components), 16, async ([id, component]) => {
							return [id, await resolveComponentStories(component, ctx.request, ctx.manifestProvider, sourceManifest.source)];
						});
						for (const [id, component] of resolved) components[id] = component;
					} catch {}
				}));
				const lists = formatMultiSourceManifestsToLists(multiSourceManifests, { withStoryIds });
				const firstSuccess = multiSourceManifests.find((m) => !m.error && !m.notice);
				if (firstSuccess) await ctx.onListAllDocumentation?.({
					context: ctx,
					manifests: {
						componentManifest: firstSuccess.componentManifest,
						docsManifest: firstSuccess.docsManifest
					},
					resultText: lists,
					sources: multiSourceManifests
				});
				return { content: [{
					type: "text",
					text: lists
				}] };
			}
			const manifests = await getManifests(ctx?.request, ctx?.manifestProvider);
			if (withStoryIds) {
				const components = manifests.componentManifest.components;
				const resolved = await mapWithConcurrency(Object.entries(components), 16, async ([id, component]) => {
					return [id, await resolveComponentStories(component, ctx?.request, ctx?.manifestProvider)];
				});
				for (const [id, component] of resolved) components[id] = component;
			}
			const lists = formatManifestsToLists(manifests, { withStoryIds });
			await ctx?.onListAllDocumentation?.({
				context: ctx,
				manifests,
				resultText: lists
			});
			return { content: [{
				type: "text",
				text: lists
			}] };
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}

//#endregion
//#region src/tools/get-documentation-for-story.ts
const GET_STORY_TOOL_NAME = "get-documentation-for-story";
const GET_STORY_DOCUMENTATION_TOOL_DESCRIPTION = "Get detailed documentation for a specific story variant of a UI component. Use this when you need to see more usage examples of a component, via the stories written for it.";
const BaseInput$1 = {
	componentId: v.string(),
	storyName: v.string()
};
function getStoryDocumentationToolSchema(options) {
	return options?.multiSource ? v.object({
		...BaseInput$1,
		...StorybookIdField
	}) : v.object(BaseInput$1);
}
function getStoryDocumentationToolMetadata(options) {
	return {
		name: GET_STORY_TOOL_NAME,
		title: "Get Documentation for Story",
		description: GET_STORY_DOCUMENTATION_TOOL_DESCRIPTION,
		schema: getStoryDocumentationToolSchema(options)
	};
}
async function addGetStoryDocumentationTool(server, enabled, options) {
	server.tool({
		...getStoryDocumentationToolMetadata(options),
		enabled
	}, async (input) => {
		try {
			const ctx = server.ctx.custom;
			const { componentId, storyName, storybookId } = input;
			const sources = ctx?.sources;
			const isMultiSource = sources && sources.some((s) => s.url);
			let source;
			if (isMultiSource) {
				if (!storybookId) return {
					content: [{
						type: "text",
						text: `storybookId is required. Available sources: ${sources.map((s) => s.id).join(", ")}. Use the ${LIST_TOOL_NAME} tool to see available sources.`
					}],
					isError: true
				};
				source = sources.find((s) => s.id === storybookId);
				if (!source) return {
					content: [{
						type: "text",
						text: `Storybook source not found: "${storybookId}". Available sources: ${sources.map((s) => s.id).join(", ")}. Use the ${LIST_TOOL_NAME} tool to see available sources.`
					}],
					isError: true
				};
			}
			let component;
			if (ctx?.resolveEntry && !source?.url) {
				const resolved = await ctx.resolveEntry(componentId, source);
				if (resolved?.kind === "component") component = resolved.component;
			} else {
				const componentEntry = (await getManifests(ctx?.request, ctx?.manifestProvider, source)).componentManifest?.components[componentId];
				if (componentEntry) component = await resolveComponentEntry(componentEntry, ctx?.request, ctx?.manifestProvider, source);
			}
			if (!component) return {
				content: [{
					type: "text",
					text: `Component not found: "${componentId}". Use the list-all-documentation tool to see available components.`
				}],
				isError: true
			};
			if (!(Array.isArray(component.stories) ? component.stories.find((s) => s.name === storyName) : void 0)) return {
				content: [{
					type: "text",
					text: `Story "${storyName}" not found for component "${componentId}". Available stories: ${Array.isArray(component.stories) ? component.stories.map((s) => s.name).join(", ") : "none"}`
				}],
				isError: true
			};
			return { content: [{
				type: "text",
				text: formatStoryDocumentation(component, storyName)
			}] };
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}

//#endregion
//#region src/tools/get-documentation.ts
const GET_TOOL_NAME = "get-documentation";
const BaseInput = { id: v.pipe(v.string(), v.description("The component or docs entry ID (e.g., \"button\")")) };
const GET_DOCUMENTATION_TOOL_DESCRIPTION = `Get documentation for a UI component or docs entry.

Returns the first ${MAX_STORIES_TO_SHOW} stories (including story IDs) with code snippets showing how props are used, plus TypeScript prop definitions. Call this before using a component to avoid hallucinating prop names, types, or valid combinations, and to answer any question about a component's props, API, or usage — reading or grepping the component source is not a substitute. Stories reveal real prop usage patterns, interactions, and edge cases that type definitions alone don't show. If the example stories don't show the prop you need, use the ${GET_STORY_TOOL_NAME} tool to fetch the story documentation for the specific story variant you need.

Example: id="button" returns Primary, Secondary, Large stories with code like <Button variant="primary" size="large"> showing actual prop combinations.`;
function getDocumentationToolSchema(options) {
	return options?.multiSource ? v.object({
		...BaseInput,
		...StorybookIdField
	}) : v.object(BaseInput);
}
function getDocumentationToolMetadata(options) {
	return {
		name: GET_TOOL_NAME,
		title: "Get Documentation",
		description: GET_DOCUMENTATION_TOOL_DESCRIPTION,
		schema: getDocumentationToolSchema(options)
	};
}
async function addGetDocumentationTool(server, enabled, options) {
	server.tool({
		...getDocumentationToolMetadata(options),
		enabled
	}, async (input) => {
		try {
			const ctx = server.ctx.custom;
			const { id, storybookId } = input;
			const sources = ctx?.sources;
			const isMultiSource = sources && sources.some((s) => s.url);
			let source;
			if (isMultiSource) {
				if (!storybookId) return {
					content: [{
						type: "text",
						text: `storybookId is required. Available sources: ${sources.map((s) => s.id).join(", ")}. Use the ${LIST_TOOL_NAME} tool to see available sources.`
					}],
					isError: true
				};
				source = sources.find((s) => s.id === storybookId);
				if (!source) return {
					content: [{
						type: "text",
						text: `Storybook source not found: "${storybookId}". Available sources: ${sources.map((s) => s.id).join(", ")}. Use the ${LIST_TOOL_NAME} tool to see available sources.`
					}],
					isError: true
				};
			}
			let component;
			let docsEntry;
			if (ctx?.resolveEntry && !source?.url) {
				const resolved = await ctx.resolveEntry(id, source);
				if (resolved?.kind === "component") component = resolved.component;
				else if (resolved?.kind === "doc") docsEntry = resolved.doc;
			} else {
				const { componentManifest, docsManifest } = await getManifests(ctx?.request, ctx?.manifestProvider, source);
				const componentEntry = componentManifest.components[id];
				const docEntry = docsManifest?.docs[id];
				if (componentEntry) component = await resolveComponentEntry(componentEntry, ctx?.request, ctx?.manifestProvider, source);
				else if (docEntry) docsEntry = await resolveDoc(docEntry, ctx?.request, ctx?.manifestProvider, source);
			}
			if (!component && !docsEntry) {
				const suffix = storybookId ? ` in source "${storybookId}"` : "";
				await ctx?.onGetDocumentation?.({
					context: ctx,
					input
				});
				return {
					content: [{
						type: "text",
						text: `Component or Docs Entry not found: "${id}"${suffix}. Use the ${LIST_TOOL_NAME} tool to see available components and documentation entries.`
					}],
					isError: true
				};
			}
			const documentation = component ?? docsEntry;
			const text = component ? formatComponentManifest(documentation) : formatDocsManifest(documentation);
			await ctx?.onGetDocumentation?.({
				context: ctx,
				input,
				foundDocumentation: documentation,
				resultText: text
			});
			return { content: [{
				type: "text",
				text
			}] };
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}

//#endregion
//#region src/index.ts
const createStorybookMcpHandler = async (options = {}) => {
	const { onSessionInitialize, ...defaultContext } = options;
	const adapter = new ValibotJsonSchemaAdapter();
	const server = new McpServer({
		name,
		version,
		description
	}, {
		adapter,
		instructions: instructions_default,
		capabilities: { tools: { listChanged: true } }
	}).withContext();
	if (onSessionInitialize) server.on("initialize", onSessionInitialize);
	await addListAllDocumentationTool(server);
	await addGetStoryDocumentationTool(server);
	await addGetDocumentationTool(server);
	const transport = new HttpTransport(server, { path: null });
	return (async (req, context) => {
		return await transport.respond(req, {
			...defaultContext,
			...context,
			request: req
		});
	});
};

//#endregion
export { COMPONENT_MANIFEST_PATH, ComponentManifestMap, DOCS_MANIFEST_PATH, DocsManifestMap, GET_STORY_TOOL_NAME, GET_TOOL_NAME, LIST_TOOL_NAME, RequiresOwnMcpError, instructions_default as STORYBOOK_MCP_INSTRUCTIONS, adaptCoreComponent, adaptCoreDoc, adaptCoreStories, addGetDocumentationTool, addGetStoryDocumentationTool, addListAllDocumentationTool, createStorybookMcpHandler, getDocumentationToolMetadata, getListAllDocumentationToolMetadata, getMultiSourceManifests, getStoryDocumentationToolMetadata, resolveComponentEntry, resolveComponentStories, resolveDoc };
