// @storybook/addon-mcp@0.7.0/dist/preset.js
import { McpServer } from "tmcp";
import { ValibotJsonSchemaAdapter } from "@tmcp/adapter-valibot";
import { HttpTransport } from "@tmcp/transport-http";
import { buffer } from "node:stream/consumers";
import { logger } from "storybook/internal/node-logger";
import { telemetry } from "storybook/internal/telemetry";
import { importModule, normalizeStoryPath } from "storybook/internal/common";
import * as v from "valibot";
import path from "node:path";
import { storyNameFromExport } from "storybook/internal/csf";
import { ComponentManifestMap, DocsManifestMap, GET_STORY_TOOL_NAME, GET_TOOL_NAME, LIST_TOOL_NAME, STORYBOOK_MCP_INSTRUCTIONS, adaptCoreComponent, adaptCoreDoc, addGetDocumentationTool, addGetStoryDocumentationTool, addListAllDocumentationTool, getDocumentationToolMetadata, getListAllDocumentationToolMetadata, getStoryDocumentationToolMetadata } from "@storybook/mcp";
import { experimental_getStatusStore } from "storybook/internal/core-server";
import { execSync } from "node:child_process";
import fs from "node:fs";
import url from "node:url";
import { stringify } from "picoquery";
import fs$1 from "node:fs/promises";

//#region package.json
var name = "@storybook/addon-mcp";
var version = "0.7.0";
var description = "Help agents automatically write and test stories for your UI components";

//#endregion
//#region src/telemetry.ts
async function collectTelemetry({ event, server, ...payload }) {
	try {
		return await telemetry("addon-mcp", {
			event,
			mcpSessionId: server.ctx.sessionId,
			clientInfo: server.ctx.sessionInfo?.clientInfo,
			clientCapabilities: server.ctx.sessionInfo?.clientCapabilities,
			...payload
		});
	} catch (error) {
		logger.debug(`Error collecting telemetry: ${String(error)}`);
	}
}

//#endregion
//#region src/utils/module-graph.ts
/**
* Access helper for Storybook's `core/module-graph` open service — the reverse index from source
* files to the story files that (transitively) import them. Replaces the removed
* `experimental_getDependencyGraphService` compatibility API.
*
* Older Storybook versions don't ship the open-service runtime; we dynamically import and treat a
* missing `getService` export (or an unregistered service) as "unsupported".
*
* The generic open-service query type (`Query`) is imported from core. The
* module-graph-specific data shapes below (`ModuleGraphStatus`, `ModuleGraphStoryHit`, etc.) are
* baked into the service's runtime definition and aren't part of core's public type surface, so we
* mirror them here; keep them in sync with `core/module-graph` if it changes.
*/
const MODULE_GRAPH_SERVICE_ID = "core/module-graph";
let probed;
async function probe() {
	if (probed !== void 0) return probed;
	try {
		const fn = (await import("storybook/internal/core-server")).getService;
		probed = typeof fn === "function" ? fn : null;
	} catch {
		probed = null;
	}
	return probed;
}
/**
* True iff the `core/module-graph` service is actually resolvable — i.e. Storybook ships the
* open-service runtime AND the service is registered in this process. Reflects registration rather
* than mere runtime presence so tool gating/badging can't drift from {@link getModuleGraphService}
* (a builder may ship the runtime but not register the service, e.g. without change detection).
*/
async function isModuleGraphSupported() {
	return await getModuleGraphService() !== void 0;
}
async function isModuleGraphSupportedByBuilder(options) {
	const builder = (await options.presets.apply("core", {}))?.builder;
	const builderName = typeof builder === "string" ? builder : builder?.name;
	if (!builderName) return false;
	try {
		return typeof (await importModule(builderName)).changeDetectionAdapter === "function";
	} catch {
		return false;
	}
}
/**
* Resolves the `core/module-graph` runtime service, or `undefined` when Storybook doesn't ship the
* open-service API or the service isn't registered (e.g. a builder without change detection, or the
* dev server isn't running). A non-undefined result doesn't mean the graph is built — await
* `queries.status.loaded(undefined)` and check for the `ready` status.
*/
async function getModuleGraphService() {
	const getService = await probe();
	if (!getService) return void 0;
	try {
		return getService(MODULE_GRAPH_SERVICE_ID);
	} catch {
		return;
	}
}

//#endregion
//#region src/utils/is-review-available.ts
const getReviewStatus = async (options, { features } = {}) => {
	const resolvedFeatures = features ?? await options.presets.apply("features", {});
	const hasFeatureFlag = !!resolvedFeatures?.experimentalReview;
	const changeDetection = !!resolvedFeatures?.changeDetection;
	return {
		available: hasFeatureFlag && changeDetection,
		availableForCli: resolvedFeatures?.experimentalReview !== false && changeDetection,
		hasFeatureFlag
	};
};

//#endregion
//#region src/tools/is-manifest-available.ts
const hasComponentManifestFeature = (features) => !!(features?.componentsManifest ?? features?.experimentalComponentsManifest);
const isDocgenServerMode = (features) => !!(features?.experimentalDocgenServer && features?.componentsManifest);
const getManifestStatus = async (options) => {
	const [features, manifests, legacyComponentManifestGenerator] = await Promise.all([
		options.presets.apply("features"),
		options.presets.apply("experimental_manifests", void 0, { manifestEntries: [] }),
		options.presets.apply("experimental_componentManifestGenerator")
	]);
	const hasFeatureFlag = hasComponentManifestFeature(features);
	const docgenServer = isDocgenServerMode(features);
	const hasManifests = docgenServer || manifests && "components" in manifests || !!legacyComponentManifestGenerator;
	return {
		available: hasFeatureFlag && hasManifests,
		hasManifests,
		hasFeatureFlag,
		docgenServer
	};
};

//#endregion
//#region src/utils/get-story-index.ts
/**
* Resolves the Storybook story index in-process via the `storyIndexGenerator` preset, rather than
* fetching `/index.json` over loopback HTTP.
*
* The dev-server registers and memoises a single `StoryIndexGenerator` (see Storybook's
* `common-preset`); applying the preset here returns that same live instance, so `getIndex()`
* hands back the up-to-date, internally-cached index without a network round-trip. This also keeps
* the index in lock-step with HMR — the generator's watcher refreshes `lastIndex` as files change.
*
* @param options - The Storybook options object carrying the resolved presets.
* @returns A promise that resolves to the StoryIndex.
* @throws If the generator preset is unavailable (e.g. not running inside a dev server).
*/
async function getStoryIndex(options) {
	const generator = await options.presets.apply("storyIndexGenerator");
	if (!generator) throw new Error("Storybook story index generator is unavailable. These MCP tools require a running Storybook dev server with a builder that exposes the story index.");
	const index = await generator.getIndex();
	logger.debug(`Story index entries found: ${Object.keys(index.entries).length}`);
	return index;
}

//#endregion
//#region src/utils/slash.ts
/**
* Normalize paths to forward slashes for cross-platform compatibility
* Storybook import paths always use forward slashes
*/
function slash(path) {
	return path.replace(/\\/g, "/");
}

//#endregion
//#region src/utils/find-story-ids.ts
function isStoryIdInput(input) {
	return "storyId" in input;
}
function normalizeImportPath(importPath) {
	return slash(normalizeStoryPath(path.posix.normalize(slash(importPath))));
}
/**
* Finds story IDs in the story index that match the given story inputs.
*
* @param index - The Storybook story index
* @param stories - Array of story inputs to search for
* @returns Array of per-input lookup results in the exact same order as the input stories
*/
function findStoryIds(index, stories) {
	const entriesList = Object.values(index.entries);
	const result = [];
	for (const storyInput of stories) {
		if (isStoryIdInput(storyInput)) {
			const foundEntry = index.entries[storyInput.storyId];
			if (foundEntry) {
				logger.debug(`Found story ID: ${foundEntry.id}`);
				result.push({
					id: foundEntry.id,
					input: storyInput
				});
			} else {
				logger.debug("No story found");
				result.push({
					input: storyInput,
					errorMessage: `No story found for story ID "${storyInput.storyId}"`
				});
			}
			continue;
		}
		const { exportName, explicitStoryName, absoluteStoryPath } = storyInput;
		const normalizedCwd = slash(process.cwd());
		const normalizedAbsolutePath = slash(absoluteStoryPath);
		const relativePath = normalizeImportPath(path.posix.relative(normalizedCwd, normalizedAbsolutePath));
		logger.debug("Searching for:");
		logger.debug({
			exportName,
			explicitStoryName,
			absoluteStoryPath,
			relativePath
		});
		const foundEntry = entriesList.find((entry) => normalizeImportPath(entry.importPath) === relativePath && [explicitStoryName, storyNameFromExport(exportName)].includes(entry.name));
		if (foundEntry) {
			logger.debug(`Found story ID: ${foundEntry.id}`);
			result.push({
				id: foundEntry.id,
				input: storyInput
			});
		} else {
			logger.debug("No story found");
			let errorMessage = `No story found for export name "${exportName}" with absolute file path "${absoluteStoryPath}"`;
			if (!explicitStoryName) errorMessage += ` (did you forget to pass the explicit story name?)`;
			result.push({
				input: storyInput,
				errorMessage
			});
		}
	}
	return result;
}

//#endregion
//#region src/utils/errors.ts
/**
* Converts an error to MCP-compatible content format
*
* @param error - The error to convert (can be any type)
* @returns A tool result with error content and isError flag
*/
const errorToMCPContent = (error) => {
	return {
		content: [{
			type: "text",
			text: `Error: ${error instanceof Error ? error.message : String(error)}`
		}],
		isError: true
	};
};

//#endregion
//#region src/tools/tool-names.ts
/**
* Tool name constants extracted to avoid circular dependencies.
*/
const PREVIEW_STORIES_TOOL_NAME = "preview-stories";
const GET_CHANGED_STORIES_TOOL_NAME = "get-changed-stories";
const GET_STORIES_BY_COMPONENT_TOOL_NAME = "get-stories-by-component";
const GET_UI_BUILDING_INSTRUCTIONS_TOOL_NAME = "get-storybook-story-instructions";
const RUN_STORY_TESTS_TOOL_NAME = "run-story-tests";
const DISPLAY_REVIEW_TOOL_NAME = "display-review";

//#endregion
//#region src/types.ts
const isLiteralEndpointPathname = (endpoint) => {
	try {
		const { pathname } = new URL(endpoint, "http://storybook.local");
		return pathname === endpoint && pathname !== "/";
	} catch {
		return false;
	}
};
const AddonOptions = v.object({
	endpoint: v.optional(v.pipe(v.string(), v.check(isLiteralEndpointPathname, "Endpoint must be a literal URL pathname"))),
	toolsets: v.optional(v.object({
		dev: v.exactOptional(v.boolean(), true),
		docs: v.exactOptional(v.boolean(), true),
		test: v.exactOptional(v.boolean(), true)
	}), {
		dev: true,
		docs: true,
		test: true
	})
});
const StoryInputProps = {
	props: v.pipe(v.optional(v.record(v.string(), v.any())), v.description(`Optional custom props to pass to the story for rendering. Use this when you don't want to render the default story,
but you want to customize some args or other props.
You can look up the component's documentation using the ${GET_UI_BUILDING_INSTRUCTIONS_TOOL_NAME} tool to see what props are available.`)),
	globals: v.pipe(v.optional(v.record(v.string(), v.any())), v.description(`Optional Storybook globals to set for the story preview. Globals are used for things like theme, locale, viewport, and other cross-cutting concerns.
Common globals include 'theme' (e.g., 'dark', 'light'), 'locale' (e.g., 'en', 'fr'), and 'backgrounds' (e.g., { value: '#000' }).`))
};
/**
* Schema for a single story input when requesting story URLs.
*/
const StoryInput = v.union([v.object({
	exportName: v.pipe(v.string(), v.description(`The export name of the story from the story file.
Use this path-based shape only when you're already editing a .stories.* file and know the export names in that file.
If you do not already have story file context, prefer the storyId shape instead of searching files.`)),
	explicitStoryName: v.pipe(v.optional(v.string()), v.description(`If the story has an explicit name set via the "name" property, that is different from the export name, provide it here.
Otherwise don't set this.`)),
	absoluteStoryPath: v.pipe(v.string(), v.description("Absolute path to the story file. Use together with exportName only when story file context is already available.")),
	...StoryInputProps
}), v.object({
	storyId: v.pipe(v.string(), v.description(`The full Storybook story ID (for example "button--primary").
Prefer this shape whenever you are not already working in a specific story file.
Use IDs discovered from ${LIST_TOOL_NAME} (withStoryIds=true) or ${GET_TOOL_NAME}.`)),
	...StoryInputProps
})]);
/**
* Schema for the array of stories to fetch URLs for.
*/
const StoryInputArray = v.array(StoryInput);

//#endregion
//#region src/tools/run-story-tests.ts
/**
* Check if addon-vitest is available by trying to import its constants.
* Returns the constants if available, undefined otherwise.
*/
async function getAddonVitestConstants() {
	try {
		const mod = await import("@storybook/addon-vitest/constants");
		return {
			TRIGGER_TEST_RUN_REQUEST: mod.TRIGGER_TEST_RUN_REQUEST,
			TRIGGER_TEST_RUN_RESPONSE: mod.TRIGGER_TEST_RUN_RESPONSE
		};
	} catch {
		return;
	}
}
const RunStoryTestsInput = v.object({
	stories: v.optional(v.pipe(StoryInputArray, v.description(`Stories to test for focused feedback. Omit this field to run tests for all available stories.
Prefer running tests for specific stories while developing to get faster feedback,
and only omit this when you explicitly need to run all tests for comprehensive verification.
Prefer { storyId } when you don't already have story file context, since this avoids filesystem discovery.
Use { storyId } when IDs were discovered from documentation tools.
Use { absoluteStoryPath + exportName } only when you're currently working in a story file and already know those values.`))),
	a11y: v.optional(v.pipe(v.boolean(), v.description("Whether to run accessibility tests. Defaults to true. Disable if you only need component test results.")), true)
});
function getRunStoryTestsToolDescription(a11yEnabled) {
	return `Run story tests.
Run them after editing anything that changes how the UI looks — components, stories, styles, CSS, themes, colors, or design tokens — shell-level substitutes like typecheck, lint, or package.json test scripts do not replace this.
Provide stories for focused runs (faster while iterating),
or omit stories to run all tests for full-project verification.
Use this continuously to monitor test results as you work on your UI components and stories.
Results will include passing/failing status` + (a11yEnabled ? `, and accessibility violation reports.
For visual/design accessibility violations (for example color contrast), ask the user before changing styles.` : ".");
}
function getRunStoryTestsToolMetadata({ a11yEnabled }) {
	return {
		name: RUN_STORY_TESTS_TOOL_NAME,
		title: "Storybook Tests",
		description: getRunStoryTestsToolDescription(a11yEnabled),
		schema: RunStoryTestsInput
	};
}
/**
* Creates a queue that ensures concurrent calls are executed in sequence.
* Call `wait()` to wait for your turn, then call the
* returned `done()` function when done to unblock the next caller.
*/
function createAsyncQueue() {
	let tail = Promise.resolve();
	/**
	* Wait for all previously queued operations to complete, then return
	* a `done` function that must be called when the current operation finishes.
	*/
	async function wait() {
		let done;
		const gate = new Promise((resolve) => {
			done = resolve;
		});
		const previousTail = tail;
		tail = previousTail.then(() => gate, () => gate);
		await previousTail.catch(() => {});
		return done;
	}
	return { wait };
}
async function addRunStoryTestsTool(server, { a11yEnabled }, enabled = () => server.ctx.custom?.toolsets?.test ?? true) {
	const addonVitestConstants = await getAddonVitestConstants();
	const testRunQueue = createAsyncQueue();
	server.tool({
		...getRunStoryTestsToolMetadata({ a11yEnabled }),
		enabled: () => {
			if (!addonVitestConstants) return false;
			return typeof enabled === "function" ? enabled() : enabled;
		}
	}, async (input) => {
		let done;
		try {
			done = await testRunQueue.wait();
			const runA11y = input.a11y ?? true;
			const { origin, options, disableTelemetry } = server.ctx.custom ?? {};
			if (!origin) throw new Error("Origin is required in addon context");
			if (!options) throw new Error("Options are required in addon context");
			const channel = options.channel;
			if (!channel) throw new Error("Channel is not available");
			let storyIds;
			let inputStoryCount = 0;
			if (input.stories) {
				const resolvedStories = findStoryIds(await getStoryIndex(options), input.stories);
				storyIds = resolvedStories.filter((story) => "id" in story).map((story) => story.id);
				inputStoryCount = input.stories.length;
				if (storyIds.length === 0) {
					const errorMessages = resolvedStories.filter((story) => "errorMessage" in story).map((story) => story.errorMessage).join("\n");
					if (!disableTelemetry) await collectTelemetry({
						event: "tool:runStoryTests",
						server,
						toolset: "test",
						runA11y,
						inputStoryCount,
						matchedStoryCount: 0,
						passingStoryCount: 0,
						failingStoryCount: 0,
						a11yViolationCount: 0,
						unhandledErrorCount: 0
					});
					return { content: [{
						type: "text",
						text: `No stories found matching the provided input.

${errorMessages}`
					}] };
				}
				logger.info(`Running focused tests for story IDs: ${storyIds.join(", ")}`);
			} else logger.info("Running tests for all stories");
			const testResults = (await triggerTestRun(channel, addonVitestConstants.TRIGGER_TEST_RUN_REQUEST, addonVitestConstants.TRIGGER_TEST_RUN_RESPONSE, storyIds, { a11y: runA11y })).result;
			if (!testResults) throw new Error("Test run response missing result data");
			const { text, summary } = formatRunStoryTestResults({
				testResults,
				runA11y,
				origin
			});
			if (!disableTelemetry) await collectTelemetry({
				event: "tool:runStoryTests",
				server,
				toolset: "test",
				runA11y,
				inputStoryCount,
				matchedStoryCount: testResults.storyIds?.length ?? storyIds?.length ?? 0,
				...summary
			});
			return { content: [{
				type: "text",
				text
			}] };
		} catch (error) {
			return errorToMCPContent(error);
		} finally {
			try {
				done?.();
			} catch (error) {
				logger.warn(`Failed to release test run queue: ${String(error)}`);
			}
		}
	});
}
/**
* Trigger a test run via Storybook channel events.
* This is the channel-based API for triggering tests in addon-vitest.
*/
function triggerTestRun(channel, triggerTestRunRequestEventName, triggerTestRunResponseEventName, storyIds, config) {
	return new Promise((resolve, reject) => {
		const requestId = `mcp-${Date.now()}`;
		let settled = false;
		const cleanup = () => {
			channel.off(triggerTestRunResponseEventName, handleResponse);
		};
		const settle = (callback) => {
			if (settled) return;
			settled = true;
			cleanup();
			callback();
		};
		const handleResponse = (payload) => {
			if (payload.requestId !== requestId) return;
			switch (payload.status) {
				case "completed":
					if (payload.result) settle(() => resolve(payload));
					else settle(() => reject(/* @__PURE__ */ new Error("Test run completed but no result was returned")));
					break;
				case "error":
					settle(() => reject(new Error(payload.error?.message ?? "Test run failed with unknown error")));
					break;
				case "cancelled":
					settle(() => reject(/* @__PURE__ */ new Error("Test run was cancelled")));
					break;
				default: settle(() => reject(/* @__PURE__ */ new Error("Unexpected test run response")));
			}
		};
		channel.on(triggerTestRunResponseEventName, handleResponse);
		const request = {
			requestId,
			actor: "addon-mcp",
			storyIds,
			config
		};
		try {
			channel.emit(triggerTestRunRequestEventName, request);
		} catch (error) {
			settle(() => reject(error instanceof Error ? error : new Error(String(error))));
		}
	});
}
function formatRunStoryTestResults({ testResults, runA11y, origin }) {
	const sections = [];
	const componentTestStatuses = testResults.componentTestStatuses;
	const passingStories = componentTestStatuses.filter((status) => status.value === "status-value:success");
	const failingStories = componentTestStatuses.filter((status) => status.value === "status-value:error");
	if (passingStories.length > 0) sections.push(formatPassingStoriesSection(passingStories));
	if (failingStories.length > 0) sections.push(formatFailingStoriesSection(failingStories));
	const a11yReports = testResults.a11yReports;
	const a11yViolationCount = runA11y ? countA11yViolations(a11yReports) : 0;
	if (runA11y && a11yReports && Object.keys(a11yReports).length > 0) {
		const a11ySection = formatA11yReportsSection({
			a11yReports,
			origin
		});
		if (a11ySection) sections.push(a11ySection);
	}
	if (testResults.unhandledErrors.length > 0) sections.push(formatUnhandledErrorsSection(testResults.unhandledErrors));
	return {
		text: sections.join("\n\n"),
		summary: {
			passingStoryCount: passingStories.length,
			failingStoryCount: failingStories.length,
			a11yViolationCount,
			unhandledErrorCount: testResults.unhandledErrors.length
		}
	};
}
function formatPassingStoriesSection(passingStories) {
	return `## Passing Stories

- ${passingStories.map((status) => status.storyId).join("\n- ")}`;
}
function formatFailingStoriesSection(statuses) {
	return `## Failing Stories

${statuses.map((status) => `### ${status.storyId}

${status.description || "No failure details available."}`).join("\n\n")}`;
}
function formatA11yReportsSection({ a11yReports, origin }) {
	const a11yViolationSections = [];
	for (const [storyId, reports] of Object.entries(a11yReports)) for (const report of reports) {
		if ("error" in report && report.error) {
			a11yViolationSections.push(`### ${storyId} - Error

${report.error.message}`);
			continue;
		}
		const violations = getA11yViolations(report);
		if (violations.length === 0) continue;
		for (const violation of violations) {
			const nodes = violation.nodes.map((node) => {
				const inspectLink = node.linkPath ? `${origin}${node.linkPath}` : void 0;
				const parts = [];
				if (node.impact) parts.push(`- **Impact**: ${node.impact}`);
				if (node.failureSummary || node.message) parts.push(`  **Message**: ${node.failureSummary || node.message}`);
				parts.push(`  **Element**: ${node.html || "(no html available)"}`);
				if (inspectLink) parts.push(`  **Inspect**: ${inspectLink}`);
				return parts.join("\n");
			}).join("\n");
			a11yViolationSections.push(`### ${storyId} - ${violation.id}

${violation.description}

#### Affected Elements
${nodes}`);
		}
	}
	if (a11yViolationSections.length === 0) return;
	return `## Accessibility Violations

${a11yViolationSections.join("\n\n")}`;
}
function formatUnhandledErrorsSection(errors) {
	return `## Unhandled Errors

${errors.map((unhandledError) => `### ${unhandledError.name || "Unknown Error"}

**Error message**: ${unhandledError.message || "No message available"}
**Path**: ${unhandledError.VITEST_TEST_PATH || "No path available"}
**Test name**: ${unhandledError.VITEST_TEST_NAME || "No test name available"}
**Stack trace**:
${unhandledError.stack || "No stack trace available"}`).join("\n\n")}`;
}
function countA11yViolations(a11yReports) {
	let count = 0;
	for (const reports of Object.values(a11yReports ?? {})) for (const report of reports) {
		if ("error" in report && report.error) continue;
		count += getA11yViolations(report).length;
	}
	return count;
}
function getA11yViolations(report) {
	if (!("violations" in report)) return [];
	const { violations } = report;
	if (!Array.isArray(violations)) return [];
	return violations.map((violation) => ({
		id: violation.id,
		description: violation.description,
		nodes: violation.nodes.map((node) => ({
			impact: typeof node.impact === "string" ? node.impact : void 0,
			failureSummary: typeof node.failureSummary === "string" ? node.failureSummary : void 0,
			html: typeof node.html === "string" ? node.html : void 0,
			linkPath: typeof node.linkPath === "string" ? node.linkPath : void 0
		}))
	}));
}

//#endregion
//#region src/utils/is-addon-a11y-enabled.ts
/**
* Check if @storybook/addon-a11y is enabled in the Storybook configuration.
*/
async function isAddonA11yEnabled(options) {
	try {
		return await options.presets.apply("isAddonA11yEnabled", false);
	} catch {
		return false;
	}
}

//#endregion
//#region src/utils/get-tool-availability.ts
/**
* Composed Storybooks with component manifests can back docs tools even when the
* local Storybook has no component manifest. Use this before feeding
* availability into the shared tool registry so live MCP registration and
* serverless AI metadata make the same docs-tool decision.
*/
function getEffectiveToolAvailability(availability, { multiSource = false } = {}) {
	if (!multiSource) return availability;
	return {
		...availability,
		docsEnabled: true,
		docsHasManifests: true,
		docsFeatureEnabled: true
	};
}
/**
* Single source of truth for the runtime gates that decide whether each tool is
* registered (and how the landing page badges it).
*
* Every dynamic gate lives here — the dependency graph, the change-detection
* pipeline, review, the component manifest (docs), addon-vitest (test) and the
* accessibility sub-feature — so the MCP server (which registers the tools) and
* the browser landing page (which shows enabled/disabled badges) can never drift
* apart. Add new gates here rather than computing them ad-hoc at a call site.
*/
async function getToolAvailability(options, { features, moduleGraphSupported: moduleGraphSupportedOverride } = {}) {
	const resolvedFeatures = features ?? await options.presets.apply("features", {});
	const [moduleGraphSupported, reviewStatus, manifestStatus, addonVitestConstants, a11yEnabled] = await Promise.all([
		moduleGraphSupportedOverride ?? isModuleGraphSupported(),
		getReviewStatus(options, { features: resolvedFeatures }),
		getManifestStatus(options),
		getAddonVitestConstants(),
		isAddonA11yEnabled(options)
	]);
	return {
		moduleGraphSupported,
		changeDetectionEnabled: resolvedFeatures?.changeDetection ?? false,
		reviewEnabled: reviewStatus.available,
		reviewEnabledForCli: reviewStatus.availableForCli,
		docsEnabled: manifestStatus.available,
		docsHasManifests: manifestStatus.hasManifests,
		docsFeatureEnabled: manifestStatus.hasFeatureFlag,
		testSupported: !!addonVitestConstants,
		a11yEnabled,
		docgenServer: manifestStatus.docgenServer
	};
}

//#endregion
//#region src/utils/estimate-tokens.ts
/**
* Checks if a character code is whitespace (space, tab, newline, carriage return)
*
* Checking char codes is slightly faster than using regex or string methods.
*/
function isWhitespace(code) {
	return code === 32 || code === 9 || code === 10 || code === 13;
}
/**
* Checks if a character code is alphanumeric or underscore
* 0-9 (48-57), A-Z (65-90), a-z (97-122), underscore (95)
*
* Checking char codes is slightly faster than using regex or string methods.
*/
function isAlphanumeric(code) {
	return code >= 48 && code <= 57 || code >= 65 && code <= 90 || code >= 97 && code <= 122 || code === 95;
}
/**
* Estimates token count from text using a fast approximation.
* Counts:
* - Continuous whitespace as a single token
* - Continuous alphanumeric sequences as single tokens
* - Each special character as an individual token
*
* This is a cheap approximation suitable for telemetry purposes.
*
* @param text - The text to estimate token count for
* @returns Estimated token count
*/
function estimateTokens(text) {
	if (!text) return 0;
	let tokenCount = 0;
	let i = 0;
	const len = text.length;
	while (i < len) {
		const code = text.charCodeAt(i);
		if (isWhitespace(code)) {
			tokenCount++;
			i++;
			while (i < len && isWhitespace(text.charCodeAt(i))) i++;
		} else if (isAlphanumeric(code)) {
			tokenCount++;
			i++;
			while (i < len && isAlphanumeric(text.charCodeAt(i))) i++;
		} else {
			tokenCount++;
			i++;
		}
	}
	return tokenCount;
}

//#endregion
//#region src/instructions/dev-instructions.md
var dev_instructions_default = "## UI Building and Story Writing Workflow\n\n- Before creating or editing components or stories, call **get-storybook-story-instructions**; its output is the source of truth for imports, story patterns, and testing conventions.\n- {{PREVIEW_STORIES_STEP}}\n- {{FINAL_LINKS_STEP}}{{DISPLAY_REVIEW_STEP}}\n- Only use story IDs returned by tools — never derive them from file names or memory. **get-stories-by-component** maps any input to stories; its description covers the workflow. No matches means no stories exist yet — say so.\n";

//#endregion
//#region src/instructions/legacy-dev-instructions.md
var legacy_dev_instructions_default = "## UI Building and Story Writing Workflow\n\n- Before creating or editing components or stories, call **get-storybook-story-instructions**.\n- Treat its output as the source of truth for imports, story patterns, and testing conventions.\n- After editing anything that changes how the UI looks — components, stories, styles, themes, colors, design tokens — call **preview-stories**, no exceptions; a shared file has no stories of its own, so preview its consumers' stories.\n- Include every returned preview URL in your final response.\n";

//#endregion
//#region src/instructions/legacy-test-instructions.md
var legacy_test_instructions_default = "## Validation Workflow\n\n- After editing anything that changes how the UI looks, run **run-story-tests** — never a package.json test script.\n- Use focused runs while iterating, then a broad pass before handoff when scope is unclear or wide.\n- Fix failing tests; never report completion while they are failing.\n";

//#endregion
//#region src/instructions/review-docs-instructions.md
var review_docs_instructions_default = "## Documentation Workflow\n\n**CRITICAL: Never hallucinate component properties!** Undocumented props do not exist — never assume them from naming or other libraries; verify every prop via these tools, not source or types in node_modules.\n\n1. Call **list-all-documentation** once at task start for component and docs IDs.\n2. Call **get-documentation** with an `id` from that list for props and usage examples.\n\nOnly reference IDs returned by these tools — never guess; scope multi-source requests with `storybookId`.\n";

//#endregion
//#region src/instructions/test-instructions.md
var test_instructions_default = "## Validation Workflow\n\n- After editing anything that changes how the UI looks, run **run-story-tests** — never a package.json test script.\n- Never report completion while story tests are failing.\n";

//#endregion
//#region src/instructions/build-server-instructions.ts
/**
* The full rule for how the agent should present links in its final
* user-facing response, delivered through the
* `get-storybook-story-instructions` output. The server instructions only
* carry a terse pointer to the same rule: MCP clients truncate server
* instructions (Claude Code cuts them at 2,048 chars), so anything beyond
* the workflow trigger must live in tool descriptions and tool results.
*
* Keyed on whether `display-review` is available in this Storybook setup.
* When available, the guidance covers both paths: ending with a review section
* after publishing, or falling back to preview URLs when no review was published
* (e.g. non-visual refactors).
*/
function getFinalLinksGuidance(reviewToolAvailable) {
	return reviewToolAvailable ? "In your final user-facing response, show one set of links — never both. If you published a review with **display-review**, finish your reply with a dedicated review section as the very last thing in the output: its own top-level heading on a line by itself (for example `## 👀 Review your changes`), then a one-line explanation that the review shows the handful of stories most relevant to this change and that, because it is AI-curated, results may be inaccurate or incomplete, then on the next line the review page as a markdown link prefixed with a 👉 so it is easy to spot, using the returned `reviewUrl` (for example `👉 [Open the Storybook review page](<reviewUrl>)`). Nothing should come after this section. Never also list the individual story or preview URLs. Avoid internal jargon like \"collection\" or \"trigger\" in anything the user reads — those are terms from this tooling, not words that mean anything to them; use plain language unless the user used the term first. A visually observable change is not finished until its review is published — never substitute preview URLs for the review. Only when there is no review because the change has no visually observable impact, say so plainly; include preview URLs only if the user asked to see specific stories." : "In your final user-facing response, include every returned preview URL so the user can verify the visual result, ordered consistently (changed-stories fallback first if relevant, then the specific preview URLs).";
}
function buildServerInstructions(options) {
	const sections = [options.docsEnabled ? "Follow these workflows when working with UI and/or Storybook. Answer questions about component props, API, or usage with the documentation tools — never from source or type definitions." : "Follow these workflows when working with UI and/or Storybook."];
	const reviewEnabled = options.reviewEnabled ?? false;
	if (options.devEnabled && !reviewEnabled) sections.push(legacy_dev_instructions_default.trim());
	else if (options.devEnabled) {
		const changeDetection = options.changeDetectionEnabled ?? false;
		const graphSupported = options.moduleGraphSupported ?? false;
		const previewStoriesStep = changeDetection ? "After editing anything that changes how the UI looks — components, stories, styles, themes, tokens — call **get-changed-stories** to discover the affected stories." : graphSupported ? "After editing anything that changes how the UI looks, call **get-stories-by-component** with the files you touched." : "After editing anything that changes how the UI looks, identify the affected stories.";
		sections.push(dev_instructions_default.replace("{{PREVIEW_STORIES_STEP}}", previewStoriesStep).replace("{{FINAL_LINKS_STEP}}", "End your final response with the review section from **display-review**'s result — never substitute preview URLs. **preview-stories** is only for mid-loop iteration or a requested direct link. If nothing visually changed, say so.").replace("{{DISPLAY_REVIEW_STEP}}", "\n- After a visually observable UI change, or when the user asks to see or browse stories/components, call **display-review** (again on each iteration) and follow its description and result. Visual work is not done until the review is published; any newly created story MUST be included.").trim());
	}
	if (options.testEnabled) sections.push((reviewEnabled ? test_instructions_default : legacy_test_instructions_default).trim());
	if (options.docsEnabled) sections.push((reviewEnabled ? review_docs_instructions_default : STORYBOOK_MCP_INSTRUCTIONS).trim());
	if (sections.length === 1) return "";
	return sections.join("\n\n");
}

//#endregion
//#region src/constants.ts
/**
* Channel event shared with Storybook core's review subsystem
* Emitted by the `display-review` tool to hand the agent's payload off to the
* core-server review channel.
*/
const PUSH_REVIEW_EVENT = "storybook/review/push-review";
/** Storybook manager route the review page is registered at. */
const REVIEW_PAGE_PATH = "/review/";
/**
* Request header the `storybook ai` CLI (and the Claude/Codex plugins built on
* it) sends on every MCP request to mark itself as a trusted local Storybook
* client. Mirrors `STORYBOOK_MCP_PROXY_HEADER` in Storybook core's
* `cli/ai/mcp/client.ts` — the two must stay in sync.
*/
const STORYBOOK_MCP_PROXY_HEADER = "X-Storybook-MCP-Proxy";
/**
* Default path the MCP server is mounted at on the Storybook dev server.
* The user can override this via the addon's `endpoint` option; everywhere
* else in the codebase that needs to compare against or fall back to the
* default should import this constant rather than hard-coding `'/mcp'`.
*/
const DEFAULT_MCP_ENDPOINT = "/mcp";

//#endregion
//#region src/utils/format-validation-issues.ts
function formatPath(path) {
	if (!path || path.length === 0) return "";
	let out = "";
	for (const segment of path) {
		const key = typeof segment === "object" && segment !== null && "key" in segment ? segment.key : segment;
		if (typeof key === "number") out += `[${key}]`;
		else if (key !== void 0 && key !== null) out += out === "" ? String(key) : `.${String(key)}`;
	}
	return out;
}
function summarizeIssue(issue) {
	const path = formatPath(issue.path);
	if (issue.kind === "schema" && (issue.type === "object" || issue.type === "strict_object" || issue.type === "loose_object") && typeof issue.expected === "string" && issue.expected.length >= 2 && issue.expected.startsWith("\"") && issue.expected.endsWith("\"") && issue.received === "undefined") {
		const field = issue.expected.slice(1, -1);
		const parentPathStr = formatPath((issue.path ?? []).slice(0, -1));
		return {
			path,
			message: `Missing required field \`${field}\`${parentPathStr ? ` at \`${parentPathStr}\`` : ""}.`
		};
	}
	return {
		path,
		message: issue.message
	};
}
/**
* Wrap a standard-schema validator so that validation failures contain
* trimmed, agent-readable issue objects instead of raw valibot metadata.
*
* Successful inputs pass through unchanged (we don't touch the parsed value).
*
* Standard Schema validators are POJOs by convention (data fields + a
* `~standard` slot, no prototype methods), so spread-and-replace is enough
* — no Proxy required.
*/
function withFriendlyErrors(schema) {
	const original = schema["~standard"];
	return {
		...schema,
		"~standard": {
			...original,
			validate(input) {
				const result = original.validate(input);
				const trim = (r) => r.issues ? { issues: r.issues.map(summarizeIssue) } : r;
				return result instanceof Promise ? result.then(trim) : trim(result);
			}
		}
	};
}

//#endregion
//#region src/tools/display-review.ts
const DISPLAY_REVIEW_TOOL_DESCRIPTION = `Publish a curated review to Storybook's review page for spot-checking **visual impact**. Each call replaces the single active review — call it again whenever the user iterates on the changes.

## When to call
- **Trigger 1 — visual change** (components, stories, CSS, themes, colors, design tokens, i18n — anything that changes how the UI looks): when the user should spot-check rendering. A shared file (token, style, util) has no stories of its own — review its consumers' stories. Skip non-visual refactors unless side-effects are plausible. Start from \`get-changed-stories\`; fall back to \`get-stories-by-component\` if change detection is unavailable. Include \`changedFiles\`.
- **Trigger 2 — browse request** ("show me the Badge component"): resolve via \`get-stories-by-component\` / \`list-all-documentation\`; you may consult other sources to interpret the ask, but IDs must still come from those tools. Pass \`changedFiles: []\` — no code changed.

## Hard rules
1. Every \`storyId\` MUST come from those tools. Reject IDs derived from file paths, story names, or memory. Unknown IDs cause a runtime error; obtain real IDs via \`get-stories-by-component\` or \`list-all-documentation\`, then retry.
2. Every story you CREATED in this change MUST appear in the review — including interaction/play-function stories. Showing the stories you modified is encouraged too. Curate by grouping, never by omission.
3. Prefer 2-5 collections; avoid one-story collections unless truly isolated.
4. Follow-up reviews: stabilize collection/story order to avoid disorientation from reshuffling.
5. Apply the field formatting rules from each schema property. Do not use em-dashes in review payload field values (title, rationale, description, etc.).
6. Do not instruct or tell the user what to do unless they explicitly ask for guidance.
7. "Collection" and "trigger" are internal terms for this tool's mechanics and mean nothing to users. Never use them in user-facing text unless the user used them first; say "group of stories" or just describe the contents in plain language.

## Curating (Trigger 1)
Trace the **visual cascade** up the **import graph** to **page-level UI surfaces** — one collection per layer (\`distance 0\` → direct importers → page context). Include **control stories** where the change is **not supposed to be visible**. **Theme tokens**, **shared styles**, and **layout primitives** need page-level coverage even from a single-file edit. **Localized changes:** affected component → **usage locations** → outer surfaces. **Larger features:** central page/module → lower-level pieces → outer **usage locations**.

## Curating (Trigger 2)
Exactly what the user asked for — **no more, no less**. Group logically or follow **story index hierarchy**.`;
/**
* Canonical schema for the agent-pushed review payload.
*
* Server-side state caching and git-branch enrichment live on the
* core server preset, not here — this tool just
* validates the agent's input and forwards it over the channel.
*/
const ReviewCollectionSchema = v.object({
	title: v.pipe(v.string(), v.description("Title describing **what** this collection consists of, phrased the way a person would say it. Avoid typographic marks and CamelCase. Plain text, no markdown.")),
	rationale: v.pipe(v.string(), v.description("Rationale explaining **why** this collection is relevant to the user. Shown alongside the title. One or two sentences. Plain text, no markdown.")),
	storyIds: v.pipe(v.array(v.string()), v.description("Story IDs that represent this collection (e.g. \"button--primary\"). The page renders exactly these."))
});
const ReviewStateSchema = v.object({
	title: v.pipe(v.string(), v.description("Terse, human-readable title for the overall review. What is this review about? Avoid typographic marks and CamelCase. Plain text, no markdown.")),
	description: v.pipe(v.string(), v.description("Description of the review scope, including what's there, why it's relevant, and what to look for. Preferably one or two sentences. At most 2 paragraphs for reviews spanning multiple topics. Markdown formatting restricted to **bold**, _italic_, and `code` (backticks). Use emphasis for the key **what** and _why_, and backticks for literal source code references like component or token names.")),
	collections: v.pipe(v.array(ReviewCollectionSchema), v.description("Groups of stories to show in the review, most relevant first. Prefer 2-5 groups.")),
	changedFiles: v.pipe(v.array(v.string()), v.description("Paths of the files you changed, most central first. Pass an empty array `[]` only when no code changed (browse requests, Trigger 2)."))
});
const DisplayReviewOutput = v.object({ reviewUrl: v.pipe(v.string(), v.description("URL of the Storybook review page. Always include this URL in your final user-facing response so the user can open it directly.")) });
function storybookRootFromRequest(request, trustedOrigin, endpoint) {
	if (!request?.url) return void 0;
	try {
		const url = new URL(request.url);
		const normalizedEndpoint = endpoint.replace(/\/$/, "");
		const normalizedPathname = url.pathname.replace(/\/$/, "");
		const rootPath = normalizedPathname.endsWith(normalizedEndpoint) ? normalizedPathname.slice(0, -normalizedEndpoint.length) : normalizedPathname.replace(/\/[^/]+$/, "");
		return `${trustedOrigin.replace(/\/$/, "")}${rootPath}`;
	} catch {
		return;
	}
}
function buildReviewUrl(ctx) {
	const trustedOrigin = ctx.origin;
	if (!trustedOrigin) throw new Error("Cannot resolve the Storybook URL: missing trusted origin in addon context.");
	return `${(ctx.request ? storybookRootFromRequest(ctx.request, trustedOrigin, ctx.endpoint ?? DEFAULT_MCP_ENDPOINT) ?? trustedOrigin : trustedOrigin).replace(/\/$/, "")}/?path=${REVIEW_PAGE_PATH}`;
}
function getDisplayReviewToolMetadata() {
	return {
		name: DISPLAY_REVIEW_TOOL_NAME,
		title: "Display Storybook review",
		description: DISPLAY_REVIEW_TOOL_DESCRIPTION,
		schema: withFriendlyErrors(ReviewStateSchema),
		outputSchema: DisplayReviewOutput
	};
}
/**
* Returns the storyIds the agent passed that don't resolve against the live
* Storybook index. Order-preserving and deduplicated so the error message
* lists each fabricated ID once, in the order the agent provided them.
*/
async function collectUnknownStoryIds(collections, options) {
	const requested = /* @__PURE__ */ new Set();
	const inOrder = [];
	for (const collection of collections) for (const id of collection.storyIds) if (!requested.has(id)) {
		requested.add(id);
		inOrder.push(id);
	}
	if (inOrder.length === 0) return [];
	const index = await getStoryIndex(options);
	return inOrder.filter((id) => !index.entries[id]);
}
function formatUnknownStoryIdsError(unknownIds) {
	const list = unknownIds.map((id) => `- \`${id}\``).join("\n");
	const plural = unknownIds.length === 1 ? "ID is" : "IDs are";
	return `Refusing to publish review: ${unknownIds.length} story ${plural} not in the live Storybook index:\n${list}\n\nThis usually means the IDs were inferred from file paths or naming conventions rather than returned by a tool. Resolve real IDs by calling \`get-stories-by-component\` (for components you've edited or want covered) or \`list-all-documentation\` (to browse the index), then retry \`display-review\` with the verified IDs. Do not invent IDs to satisfy this check.`;
}
async function addDisplayReviewTool(server, enabled = () => server.ctx.custom?.toolsets?.dev ?? true) {
	server.tool({
		...getDisplayReviewToolMetadata(),
		enabled
	}, async (input) => {
		try {
			const customContext = server.ctx.custom;
			if (!customContext?.origin) throw new Error("Cannot resolve the Storybook URL: missing trusted origin in addon context.");
			if (!customContext.options) throw new Error("Storybook options are required in addon context.");
			const unknownIds = await collectUnknownStoryIds(input.collections, customContext.options);
			if (unknownIds.length > 0) throw new Error(formatUnknownStoryIdsError(unknownIds));
			const reviewUrl = buildReviewUrl({
				origin: customContext.origin,
				request: customContext.request,
				endpoint: customContext.endpoint
			});
			server.ctx.custom?.options?.channel?.emit(PUSH_REVIEW_EVENT, input);
			const collectionCount = input.collections.length;
			const storyCount = input.collections.reduce((n, c) => n + c.storyIds.length, 0);
			if (!customContext.disableTelemetry) await collectTelemetry({
				event: "tool:displayReview",
				server,
				toolset: "dev",
				collectionCount,
				storyCount,
				changedFileCount: input.changedFiles.length
			});
			return {
				content: [{
					type: "text",
					text: `Review applied: ${collectionCount} collection${collectionCount === 1 ? "" : "s"}, ${storyCount} stor${storyCount === 1 ? "y" : "ies"}. Storybook is already running at ${customContext.origin} — reuse it. Do NOT start another Storybook or change its port to view this review; the running instance already serves it.

Two things you must do now, both of them:
1. **Open ${reviewUrl} yourself in your preview browser.** If you have any browser-preview or navigate tool in this session (e.g. preview_eval or an equivalent), call it on this URL so the review opens in your preview window immediately. Don't merely print the link and stop — actually open it.
2. **Show the link to the user too.** End your final response with a dedicated review section as the very last thing: its own heading on a line by itself (e.g. \`## 👀 Review your changes\`), then a one-line explanation of what the review is, then on the next line the review page as a markdown link prefixed with a 👉 so it's easy to spot: \`👉 [Open the Storybook review page](${reviewUrl})\`. For the explanation, use something like: "The review shows the ${storyCount} stor${storyCount === 1 ? "y" : "ies"} most relevant for you to review right now. Because this is AI-curated, results may be inaccurate or incomplete." Put nothing after the link — not a trailing sentence the user has to hunt for. The user needs to see this link even after you've opened it yourself.`
				}],
				structuredContent: { reviewUrl }
			};
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}

//#endregion
//#region src/utils/detect-unreachable-changes.ts
const SOURCE_EXT_RE = /\.(?:tsx?|jsx?|mjs|cjs)$/i;
/**
* Module-graph query batch size. Bounds the per-call file list so a large working tree doesn't
* issue one oversized reverse-index query; we early-exit between chunks once `maxFiles` is reached.
*/
const CHUNK_SIZE = 50;
/**
* git status --porcelain produces lines like `XY filename`, where XY are two
* status characters (M, A, ?, etc). The pattern is rigid; we slice the prefix.
* Rename lines are formatted as `R  old -> new` — keep only the new path.
*/
function parsePorcelain(output) {
	return output.split("\n").filter(Boolean).map((line) => {
		const rest = line.slice(3);
		const arrow = rest.indexOf(" -> ");
		return arrow >= 0 ? rest.slice(arrow + 4).trim() : rest.trim();
	});
}
/**
* Lists working-tree modified files that are NOT reached from any story file
* via Storybook's reverse dependency graph. This is the "your edit isn't in
* the graph; you'll need to grep" case — typical for theme tokens, decorator
* config, and other infrastructure files consumed via Storybook's preview
* runtime rather than story-file imports.
*
* Returns workspace-relative paths, suitable for inlining into a tool
* response. Empty when:
*   - change detection is inactive,
*   - the working tree is clean,
*   - or every modified file IS in the graph (the caller should use the
*     status-store result directly).
*/
async function detectUnreachableChanges(maxFiles = 10) {
	const moduleGraph = await getModuleGraphService();
	if (!moduleGraph) return [];
	if ((await moduleGraph.queries.status.loaded(void 0)).value !== "ready") return [];
	const workingDir = process.cwd();
	let porcelain;
	try {
		porcelain = execSync("git status --porcelain", {
			cwd: workingDir,
			encoding: "utf-8",
			stdio: [
				"ignore",
				"pipe",
				"ignore"
			]
		});
	} catch {
		return [];
	}
	const relFiles = parsePorcelain(porcelain).filter((f) => SOURCE_EXT_RE.test(f));
	if (relFiles.length === 0) return [];
	const unreachable = [];
	for (let start = 0; start < relFiles.length && unreachable.length < maxFiles; start += CHUNK_SIZE) {
		const chunk = relFiles.slice(start, start + CHUNK_SIZE);
		const absChunk = chunk.map((rel) => slash(path.resolve(workingDir, rel)));
		const hits = await moduleGraph.queries.storiesForFiles.loaded({ files: absChunk });
		for (const [i, rel] of chunk.entries()) {
			if (unreachable.length >= maxFiles) break;
			if ((hits[i]?.length ?? 0) === 0) unreachable.push(rel);
		}
	}
	return unreachable;
}
/**
* Maximum unreachable files listed inline in the short header banner. Beyond
* this we summarise the overflow as `+N more` to keep the banner one-line-ish.
*/
const BANNER_INLINE_LIMIT = 3;
/**
* Short "front-loaded" version of {@link formatPartialCoverageHint}, designed
* to survive truncation and compaction. The full hint still trails the
* response body unchanged — this banner is purely a salience aid for the
* non-empty case where 1000+ story bullets would otherwise bury the tail.
*
* Empty when no unreachable files were detected.
*/
function formatPartialCoverageBanner(unreachable) {
	if (unreachable.length === 0) return "";
	const fileList = unreachable.length <= BANNER_INLINE_LIMIT ? unreachable.join(", ") : `${unreachable.slice(0, BANNER_INLINE_LIMIT).join(", ")}, +${unreachable.length - BANNER_INLINE_LIMIT} more`;
	const noun = unreachable.length === 1 ? "file" : "files";
	return `⚠ Coverage gap: ${unreachable.length} modified ${noun} unreachable from any story (${fileList}) — full sanity-check note at end of this response.\n\n`;
}
/**
* Formats the unreachable-files list into a hint appended to
* `get-changed-stories`' empty response. Empty string when no hint applies.
*/
function formatUnreachableHint(unreachable) {
	if (unreachable.length === 0) return "";
	return `\n\nThe following working-tree file(s) are modified but unreachable from any story (no static import path connects them — they are likely theme tokens, decorators, or other Storybook-preview-runtime files):\n${unreachable.map((f) => `- ${f}`).join("\n")}\n\nFor these, grep the codebase for their exports (e.g. specific tokens or symbols) to find runtime consumers, then call \`get-stories-by-component\` with those consumer file paths.`;
}
/**
* Formats a sanity-check hint for the *non-empty* response case: there ARE
* changed stories, but the working tree also contains modified files that
* aren't reachable from any story (typically because the diff bundles a prior
* sub-change on a component with a follow-up edit to shared infrastructure).
* In that situation, the changed-stories list is real but incomplete w.r.t.
* the agent's most recent sub-edit — the agent has to actively check coverage
* rather than trust the list.
*
* Empty string when no hint applies.
*/
function formatPartialCoverageHint(unreachable) {
	if (unreachable.length === 0) return "";
	return `\n\nCoverage sanity check: the working tree also contains modified file(s) that aren't reachable from any story above (no static import path connects them — typically theme tokens, decorators, or other preview-runtime files):\n${unreachable.map((f) => `- ${f}`).join("\n")}\n\nThe list above is real but may be stale w.r.t. these files — they're often left over from an earlier sub-change in the same diff. Before composing a review, grep the codebase for their exports and call \`get-stories-by-component\` with the runtime consumers' file paths. Do not assume the list above already covers them, and never invent story IDs to fill the gap.`;
}

//#endregion
//#region src/tools/get-changed-stories.ts
const GET_CHANGED_STORIES_TOOL_DESCRIPTION = `Get Storybook stories marked as new, modified, or related. Returns story metadata only (no URLs).

The result reflects the cumulative working-tree diff, not just your latest edit — after multiple edits in one session, a non-empty result may cover an earlier sub-change and miss your most recent one. Check that every file you touched is represented; for any that isn't, find its consumer components and pass their paths to get-stories-by-component instead. The response surfaces this gap with a "coverage sanity check" hint when it detects unreachable working-tree files.`;
const CHANGE_DETECTION_TYPE = "storybook/change-detection";
const INCLUDED_STATUS_VALUES = new Set([
	"status-value:new",
	"status-value:modified",
	"status-value:affected"
]);
function statusPriority(statusValue) {
	if (statusValue === "status-value:new") return 0;
	if (statusValue === "status-value:modified") return 1;
	return 2;
}
function getChangedStoriesToolMetadata() {
	return {
		name: GET_CHANGED_STORIES_TOOL_NAME,
		title: "Get changed stories metadata",
		description: GET_CHANGED_STORIES_TOOL_DESCRIPTION
	};
}
async function addGetChangedStoriesTool(server, enabled = () => server.ctx.custom?.toolsets?.dev ?? true, { reviewEnabled = false } = {}) {
	server.tool({
		...getChangedStoriesToolMetadata(),
		enabled
	}, async () => {
		try {
			const { options, disableTelemetry } = server.ctx.custom ?? {};
			if (!options) throw new Error("Storybook options are required in addon context");
			const allStatuses = experimental_getStatusStore(CHANGE_DETECTION_TYPE).getAll();
			const changedStoriesFromStatusStore = [];
			for (const byType of Object.values(allStatuses)) {
				const status = byType?.[CHANGE_DETECTION_TYPE];
				if (status?.value && INCLUDED_STATUS_VALUES.has(status.value)) changedStoriesFromStatusStore.push(status);
			}
			if (changedStoriesFromStatusStore.length === 0) {
				if (!disableTelemetry) await collectTelemetry({
					event: "tool:getChangedStories",
					server,
					toolset: "dev",
					storyCount: 0,
					newStoryCount: 0,
					modifiedStoryCount: 0,
					affectedStoryCount: 0
				});
				return { content: [{
					type: "text",
					text: `No new, modified, or related stories detected.${formatUnreachableHint(await detectUnreachableChanges())}`
				}] };
			}
			const index = await getStoryIndex(options);
			const stories = changedStoriesFromStatusStore.flatMap(({ storyId, value }) => {
				const entry = index.entries[storyId];
				if (!entry) return [];
				return [{
					storyId,
					statusValue: value,
					title: entry.title,
					name: entry.name,
					importPath: entry.importPath
				}];
			});
			stories.sort((a, b) => {
				const priorityDelta = statusPriority(a.statusValue) - statusPriority(b.statusValue);
				return priorityDelta !== 0 ? priorityDelta : a.storyId.localeCompare(b.storyId);
			});
			const buckets = {
				new: stories.filter((story) => story.statusValue === "status-value:new"),
				modified: stories.filter((story) => story.statusValue === "status-value:modified"),
				affected: stories.filter((story) => story.statusValue === "status-value:affected")
			};
			const counts = {
				new: buckets.new.length,
				modified: buckets.modified.length,
				affected: buckets.affected.length
			};
			if (!disableTelemetry) await collectTelemetry({
				event: "tool:getChangedStories",
				server,
				toolset: "dev",
				storyCount: stories.length,
				newStoryCount: counts.new,
				modifiedStoryCount: counts.modified,
				affectedStoryCount: counts.affected
			});
			const unreachable = await detectUnreachableChanges();
			let text = `${formatPartialCoverageBanner(unreachable)}Detected ${stories.length} changed stor${stories.length === 1 ? "y" : "ies"} (${counts.new} new, ${counts.modified} modified, ${counts.affected} related).`;
			if (reviewEnabled && stories.length > 0) text += `\n\nNext: if the change is visually observable, publish the review now — call **display-review** curating these story IDs. That review link is how you finish; do not substitute individual preview URLs for it.`;
			const serializeStory = ({ storyId, title, name, importPath }) => `- \`${storyId}\`: ${title} / ${name} (\`${importPath}\`)`;
			if (buckets.new.length > 0) {
				text += `\n\nNew stories:\n`;
				text += buckets.new.map(serializeStory).join("\n");
			}
			if (buckets.modified.length > 0) {
				text += `\n\nModified stories:\n`;
				text += buckets.modified.map(serializeStory).join("\n");
			}
			if (buckets.affected.length > 0) {
				text += `\n\nRelated stories:\n`;
				text += buckets.affected.map(serializeStory).join("\n");
			}
			text += formatPartialCoverageHint(unreachable);
			return { content: [{
				type: "text",
				text
			}] };
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}

//#endregion
//#region src/utils/resolve-component-stories.ts
const SOURCE_EXTENSIONS = [
	".ts",
	".tsx",
	".js",
	".jsx",
	".mjs",
	".cjs"
];
const INDEX_BASENAMES = SOURCE_EXTENSIONS.map((ext) => `index${ext}`);
/**
* Expands `Foo/Foo.tsx` ↔ `Foo/index.tsx` so a query against either form hits stories that
* imported the other. Only expands when filename or dirname clearly names the directory's
* main export.
*/
function expandBarrelTargets(absoluteComponentPath) {
	const targets = new Set([absoluteComponentPath]);
	const dir = path.dirname(absoluteComponentPath);
	const ext = path.extname(absoluteComponentPath);
	const base = path.basename(absoluteComponentPath, ext);
	const dirName = path.basename(dir);
	if (base.toLowerCase() === "index") for (const candidateExt of SOURCE_EXTENSIONS) {
		const candidate = path.join(dir, `${dirName}${candidateExt}`);
		if (fs.existsSync(candidate)) targets.add(candidate);
	}
	else if (base.toLowerCase() === dirName.toLowerCase()) for (const indexBasename of INDEX_BASENAMES) {
		const candidate = path.join(dir, indexBasename);
		if (fs.existsSync(candidate)) targets.add(candidate);
	}
	return [...targets].map((p) => slash(path.normalize(p)));
}
/**
* Canonicalises a path via `realpath`. Fixes silent misses on case-insensitive filesystems
* (macOS APFS) where `BUTTON/Button.tsx` and `Button/Button.tsx` resolve to the same file but
* only the canonical form is in the reverse-index keyset.
*
* Returns `undefined` if the path doesn't exist on disk — the caller surfaces that as a
* distinct "path not found" outcome rather than a generic "no stories".
*/
function canonicalise(absolutePath) {
	try {
		return fs.realpathSync.native(absolutePath);
	} catch (err) {
		if (err.code === "ENOENT") return void 0;
		throw err;
	}
}
/**
* Normalizes a story-index `importPath` to the `./`-prefixed, forward-slash form the module graph
* returns for its story files (e.g. `./src/Button.stories.tsx`). The module graph keys story files
* this way, so we match on it instead of the absolute path.
*/
function toStoryIndexPath(importPath) {
	const normalized = slash(importPath);
	if (normalized.startsWith("./") || normalized.startsWith("../")) return normalized;
	return `./${normalized}`;
}
/**
* Map story-index-relative story-file path → story IDs declared in that file. Skips virtual entries.
* Keys match the `storyFile` values returned by the module graph's `storiesForFiles`.
*/
function buildStoryIdsByFile(storyIndex) {
	const storyIdsByFile = /* @__PURE__ */ new Map();
	for (const entry of Object.values(storyIndex.entries)) {
		if (entry.type !== "story" || entry.importPath.startsWith("virtual:")) continue;
		const filePath = toStoryIndexPath(entry.importPath);
		let ids = storyIdsByFile.get(filePath);
		if (!ids) {
			ids = /* @__PURE__ */ new Set();
			storyIdsByFile.set(filePath, ids);
		}
		ids.add(entry.id);
	}
	return storyIdsByFile;
}
/** Builds the unavailable reason for a non-`ready` module-graph status. */
function reasonForStatus(status) {
	switch (status.value) {
		case "booting": return "Storybook's story module graph hasn't built yet — it is still being constructed. Retry shortly.";
		case "unavailable": return `Storybook's story module graph is unavailable: ${status.reason}`;
		case "error": return `Storybook's story module graph failed to build: ${status.error.message}`;
		case "ready": return "Storybook's story module graph is ready.";
		default: return status;
	}
}
/**
* Looks up stories that consume each component path via Storybook's `core/module-graph` reverse
* index. Returns `{ available: false }` when the graph isn't ready; otherwise per-path results.
*/
async function resolveComponentStories(request, deps) {
	const moduleGraph = await getModuleGraphService();
	if (!moduleGraph) return {
		available: false,
		reason: "Storybook's story module graph is unavailable. This Storybook version may not ship the open-service API, the module-graph service isn't registered (e.g. a builder without change detection), or the dev server is not running."
	};
	const status = await moduleGraph.queries.status.loaded(void 0);
	if (status.value !== "ready") return {
		available: false,
		reason: reasonForStatus(status)
	};
	const workingDir = deps.workingDir ?? process.cwd();
	const storyIdsByFile = buildStoryIdsByFile(await deps.getStoryIndex());
	const resolved = [...new Set(request.componentPaths)].map((componentPath) => {
		const absolute = path.resolve(workingDir, componentPath);
		const canonical = canonicalise(absolute);
		const echo = path.normalize(componentPath);
		if (!canonical) return {
			componentPath: echo,
			targets: [],
			pathNotFound: true
		};
		const targets = new Set(expandBarrelTargets(absolute));
		if (canonical !== absolute) for (const t of expandBarrelTargets(canonical)) targets.add(t);
		return {
			componentPath: echo,
			targets: [...targets]
		};
	});
	const allTargets = [...new Set(resolved.flatMap((c) => c.targets))];
	const hitsByTarget = /* @__PURE__ */ new Map();
	if (allTargets.length > 0) {
		const batched = await moduleGraph.queries.storiesForFiles.loaded({ files: allTargets });
		allTargets.forEach((target, i) => {
			hitsByTarget.set(target, batched[i] ?? []);
		});
	}
	return {
		available: true,
		results: resolved.map((component) => {
			if (component.pathNotFound) return {
				componentPath: component.componentPath,
				matches: [],
				pathNotFound: true
			};
			const byStoryId = /* @__PURE__ */ new Map();
			for (const target of component.targets) {
				const hits = hitsByTarget.get(target) ?? [];
				for (const { storyFile, depth } of hits) {
					const storyIds = storyIdsByFile.get(storyFile);
					if (!storyIds) continue;
					for (const storyId of storyIds) {
						const existing = byStoryId.get(storyId);
						if (existing === void 0 || depth < existing) byStoryId.set(storyId, depth);
					}
				}
			}
			const matches = [...byStoryId.entries()].map(([storyId, depth]) => ({
				storyId,
				depth
			})).sort((a, b) => {
				if (a.depth !== b.depth) return a.depth - b.depth;
				return a.storyId.localeCompare(b.storyId);
			});
			return {
				componentPath: component.componentPath,
				matches
			};
		})
	};
}

//#endregion
//#region src/tools/get-stories-by-component.ts
/** When omitted by the caller, applied internally to keep result sets actionable on real codebases. */
const DEFAULT_MAX_DISTANCE = 3;
/**
* Review-conditional like the preview-stories description: with the
* `experimentalReview` flag off, display-review is not registered, so the
* review-off variant must not point agents at a tool that doesn't exist.
*/
function getStoriesByComponentToolDescription({ reviewEnabled = false } = {}) {
	return `Map component source files to the stories that render them, returning grounded \`storyId\` values from the live Storybook index — hand these to ${reviewEnabled ? `${PREVIEW_STORIES_TOOL_NAME} or display-review` : PREVIEW_STORIES_TOOL_NAME} instead of guessing.

Reach for this whenever you need story IDs, whatever shape the input has: ${reviewEnabled ? `files you just edited, a feature/domain/topic the user named, a query like "all consumers of X", or an autonomous review after a UI change` : `files you just edited, a feature/domain/topic the user named, or a query like "all consumers of X"`}. First resolve the input to a list of absolute component file paths using filesystem search (grep / Glob / find) and code reading — that bridge is yours to build; this tool starts where it ends. One common trap: when the changed file is _shared_ infrastructure (theme token, design token, util, hook, CSS module) it isn't itself a component — grep for its consumers and pass _their_ paths, not the shared file's. If the symbol you grepped looks like one member of a related group (sibling tokens, neighboring exports), widen to the rest of the group too — a too-narrow grep silently drops stories. Try \`${GET_CHANGED_STORIES_TOOL_NAME}\` first for "I just edited X" when it's available; if a file you touched is missing from its response, treat that file as the shared-infrastructure case and route its consumers through this tool.

Results are sorted by \`distance\` (0 = the path you passed is itself a story file, 1 = direct importer, 2+ = transitive; lower = stronger). Shared primitives are usually consumed through wrapper components, so the distance-1 bucket is often empty — the default \`maxDistance: 3\` keeps that cascade visible while capping noise from wide decorators; raise it to widen recall, lower it to tighten precision. ${reviewEnabled ? `For display-review, the distance buckets map onto the visual cascade (the component itself → direct importers → page-level context) — one collection per layer; when several stories of a component share a distance, prefer the variant whose name signals it renders the changed surface.` : `The distance buckets map onto the visual cascade (the component itself → direct importers → page-level context) — use them to decide which stories to preview; when several stories of a component share a distance, prefer the variant whose name signals it renders the changed surface.`}

Never invent IDs from file names, feature names, or memory; title strings can be overridden by story authors, so only IDs returned by discovery tools resolve. If a component has no matches here, it has no stories yet (say so, don't fabricate).

Backed by Storybook's live reverse dependency graph, available only when the dev server runs a builder that supports change detection (e.g. Vite) — otherwise returns a typed error.`;
}
const GetStoriesByComponentInput = v.object({
	componentPaths: v.pipe(v.array(v.string()), v.minLength(1), v.description(`Absolute paths to component source files (e.g. "/repo/src/Button.tsx").
Pass the components you actually want stories for — typically files you just read, edited, or that the user mentioned.
Relative paths are also accepted and resolved against the Storybook working directory, but absolute paths are preferred for unambiguous results.
Story files (\`*.stories.*\`) are accepted too: they appear at distance 0 as self-matches, plus any reverse-graph hits (other stories that import them).`)),
	maxDistance: v.pipe(v.optional(v.pipe(v.number(), v.minValue(1), v.integer())), v.description(`Ceiling on the import depth to include in results. Must be a positive integer.
- 1: only stories that directly import the component.
- 2+: also include stories that reach the component through N hops.
Defaults to ${DEFAULT_MAX_DISTANCE}; raise it to widen recall, lower it to tighten precision. Shared components (Button, Icon, …) accumulate noisy indirect matches at distance ≥ 3, so the default cap protects against runaway results.`))
});
const StoryMatch = v.object({
	storyId: v.string(),
	title: v.string(),
	name: v.string(),
	importPath: v.string(),
	distance: v.pipe(v.number(), v.description("Import-graph depth from the story file to the component (lower = stronger). 0: the path you passed is itself a story file (self-match). 1: story file directly imports the component. 2+: reached through N hops."))
});
const ClippedByMaxDistanceSchema = v.pipe(v.object({
	count: v.number(),
	distances: v.array(v.number())
}), v.description("Present only when `maxDistance` filtered out one or more matches. `count` is how many were dropped; `distances` lists the (sorted, distinct) distances those dropped matches sat at — widen `maxDistance` to include them."));
const GetStoriesByComponentOutput = v.object({ results: v.array(v.object({
	componentPath: v.string(),
	matches: v.array(StoryMatch),
	clipped: v.optional(ClippedByMaxDistanceSchema),
	pathNotFound: v.pipe(v.optional(v.boolean()), v.description("`true` when no file exists at the resolved absolute path. Distinguishes a typo from \"this component has no stories yet\". The agent should re-check the path it sent."))
})) });
function serializeMatch(match) {
	return `  - \`${match.storyId}\`: ${match.title} / ${match.name} (\`${match.importPath}\`)`;
}
function pluralize(n, singular, plural = `${singular}s`) {
	return n === 1 ? singular : plural;
}
function formatClippedTail(clipped, maxDistance) {
	const dists = clipped.distances;
	const rangeText = dists.length === 1 ? `distance ${dists[0]}` : `distances ${dists[0]}..${dists[dists.length - 1]}`;
	return `+${clipped.count} more ${pluralize(clipped.count, "story", "stories")} at ${rangeText} hidden by \`maxDistance: ${maxDistance}\``;
}
function serializeComponentSection(componentPath, matches, options = {}) {
	const { maxDistance, clipped, pathNotFound } = options;
	if (pathNotFound) return `${componentPath}: path does not exist on disk — re-check the path you sent.`;
	if (matches.length === 0) {
		if (clipped && clipped.count > 0 && maxDistance !== void 0) return `${componentPath}: no stories within \`maxDistance: ${maxDistance}\` — ${formatClippedTail(clipped, maxDistance)}.`;
		return `${componentPath}: no stories found`;
	}
	const byDistance = /* @__PURE__ */ new Map();
	for (const m of matches) {
		const bucket = byDistance.get(m.distance) ?? [];
		bucket.push(m);
		byDistance.set(m.distance, bucket);
	}
	const distances = [...byDistance.keys()].sort((a, b) => a - b);
	const minDist = distances[0];
	const maxDist = distances[distances.length - 1];
	const componentCount = new Set(matches.map((m) => m.title)).size;
	const bucketSummary = distances.map((d) => `d${d}=${byDistance.get(d).length}`).join(", ");
	const summary = `→ ${matches.length} ${pluralize(matches.length, "story", "stories")} across ${componentCount} ${pluralize(componentCount, "component")}, distances ${minDist}..${maxDist} (${bucketSummary})`;
	const lines = [`${componentPath}:`, summary];
	for (const d of distances) {
		lines.push(`distance ${d}:`);
		for (const m of byDistance.get(d)) lines.push(serializeMatch(m));
	}
	if (clipped && clipped.count > 0 && maxDistance !== void 0) lines.push(`  (${formatClippedTail(clipped, maxDistance)}.)`);
	return lines.join("\n");
}
function applyMaxDistance(depths, maxDistance) {
	if (maxDistance === void 0) return { kept: depths };
	const kept = [];
	const clippedDistances = /* @__PURE__ */ new Set();
	let clippedCount = 0;
	for (const d of depths) if (d.depth <= maxDistance) kept.push(d);
	else {
		clippedCount++;
		clippedDistances.add(d.depth);
	}
	return {
		kept,
		clipped: clippedCount > 0 ? {
			count: clippedCount,
			distances: [...clippedDistances].sort((a, b) => a - b)
		} : void 0
	};
}
function getStoriesByComponentToolMetadata({ reviewEnabled = false } = {}) {
	return {
		name: GET_STORIES_BY_COMPONENT_TOOL_NAME,
		title: "Get stories for component files",
		description: getStoriesByComponentToolDescription({ reviewEnabled }),
		schema: GetStoriesByComponentInput,
		outputSchema: GetStoriesByComponentOutput
	};
}
async function addGetStoriesByComponentTool(server, enabled = () => server.ctx.custom?.toolsets?.dev ?? true, { reviewEnabled = false } = {}) {
	server.tool({
		...getStoriesByComponentToolMetadata({ reviewEnabled }),
		enabled
	}, async (input) => {
		try {
			const { options, disableTelemetry } = server.ctx.custom ?? {};
			if (!options) throw new Error("Storybook options are required in addon context");
			const index = await getStoryIndex(options);
			const lookup = await resolveComponentStories({ componentPaths: input.componentPaths }, { getStoryIndex: async () => index });
			if (!lookup.available) return {
				content: [{
					type: "text",
					text: lookup.reason ?? "Storybook's story dependency graph is unavailable. Make sure the dev server is running with a builder that supports change detection."
				}],
				isError: true
			};
			const effectiveMaxDistance = input.maxDistance ?? DEFAULT_MAX_DISTANCE;
			const results = (lookup.results ?? []).map((entry) => {
				if (entry.pathNotFound) return {
					componentPath: entry.componentPath,
					matches: [],
					pathNotFound: true
				};
				const { kept, clipped } = applyMaxDistance(entry.matches, effectiveMaxDistance);
				const matches = [];
				for (const { storyId, depth } of kept) {
					const indexEntry = index.entries[storyId];
					if (!indexEntry || indexEntry.type !== "story") continue;
					matches.push({
						storyId: indexEntry.id,
						title: indexEntry.title,
						name: indexEntry.name,
						importPath: indexEntry.importPath,
						distance: depth
					});
				}
				return {
					componentPath: entry.componentPath,
					matches,
					clipped
				};
			});
			const totalMatches = results.reduce((sum, r) => sum + r.matches.length, 0);
			const unmatchedCount = results.filter((r) => !r.pathNotFound && r.matches.length === 0).length;
			const textSections = results.map(({ componentPath, matches, clipped, pathNotFound }) => serializeComponentSection(componentPath, matches, {
				maxDistance: effectiveMaxDistance,
				clipped,
				pathNotFound
			}));
			const text = textSections.length > 0 ? textSections.join("\n\n") : "No component paths provided.";
			if (!disableTelemetry) await collectTelemetry({
				event: "tool:getStoriesByComponent",
				server,
				toolset: "dev",
				componentCount: input.componentPaths.length,
				matchedComponentCount: input.componentPaths.length - unmatchedCount,
				totalMatchCount: totalMatches,
				maxDistance: effectiveMaxDistance
			});
			return {
				content: [{
					type: "text",
					text
				}],
				structuredContent: { results }
			};
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}

//#endregion
//#region src/instructions/storybook-story-instructions.md
var storybook_story_instructions_default = "# Writing User Interfaces\n\nWhen writing UI, prefer breaking larger components up into smaller parts.\n\nALWAYS write a Storybook story for any component written. If editing a component, ensure appropriate changes have been made to stories for that component.\n{{DOCS_WORKFLOW_GUIDANCE}}\n\n## How to write good stories\n\nGoal: Cover every distinct piece of business logic and state the component can reach (happy paths, error/edge states, loading, permissions/roles, empty states, variations from props/context). Avoid redundant stories that show the same logic.\n\nInteractivity: If the component is interactive, add Interaction tests using play functions that drive the UI with storybook/test utilities (e.g., fn, userEvent, expect). Simulate key user flows: clicking buttons/links, typing, focus/blur, keyboard nav, form submit, async responses, toggle/selection changes, pagination/filters, etc. When passing `fn` functions as `args` for callback functions, make sure to add a play function which interacts with the component and assert whether the callback function was actually called.\n\nData/setup: Provide realistic props, state, and mocked data. Include meaningful labels/text to make behaviors observable. Stub network/services with deterministic fixtures; keep stories reliable.\n\nAssertions: In play functions, assert the visible outcome of the interaction (text, aria state, enabled/disabled, class/state changes, emitted events). Prefer role/label-based queries.\n\nVariants to consider (pick only those that change behavior): default vs. alternate themes; loading vs. loaded vs. empty vs. error; validated vs. invalid input; permissions/roles/capabilities; feature flags; size/density/layout variants that alter logic.\n\nAccessibility: Use semantic roles/labels; ensure focusable/keyboard interactions are test-covered where relevant.\n\nNaming/structure: Use clear story names that describe the scenario (“Error state after failed submit”). Group related variants logically; don’t duplicate.\n\nImports/format: Import Meta/StoryObj from the framework package; import test helpers from storybook/test (not @storybook/test). Keep stories minimal—only what's needed to demonstrate behavior.\n\n## Storybook 9 Essential Changes for Story Writing\n\n### Package Consolidation\n\n#### `Meta` and `StoryObj` imports\n\nUpdate story imports to use the framework package:\n\n```diff\n- import { Meta, StoryObj } from '{{RENDERER}}';\n+ import { Meta, StoryObj } from '{{FRAMEWORK}}';\n```\n\n#### Test utility imports\n\nUpdate test imports to use `storybook/test` instead of `@storybook/test`\n\n```diff\n- import { fn } from '@storybook/test';\n+ import { fn } from 'storybook/test';\n```\n\n### Global State Changes\n\nThe `globals` annotation has be renamed to `initialGlobals`:\n\n```diff\n// .storybook/preview.js\nexport default {\n- globals: { theme: 'light' }\n+ initialGlobals: { theme: 'light' }\n};\n```\n\n### Autodocs Configuration\n\nInstead of `parameters.docs.autodocs` in main.js, use tags:\n\n```js\n// .storybook/preview.js or in individual stories\nexport default {\n	tags: ['autodocs'], // generates autodocs for all stories\n};\n```\n\n### Mocking imports in Storybook\n\nTo mock imports in Storybook, use Storybook's mocking features. ALWAYS mock external dependencies to ensure stories render consistently.\n\n1. **Register in the mock in Storybook's preview file**:\n   To mock dependendencies, you MUST register a module mock in `.storybook/preview.ts` (or equivalent):\n\n```js\nimport { sb } from 'storybook/test';\n\n// Prefer spy mocks (keeps functions, but allows to override them and spy on them)\nsb.mock(import('some-library'), { spy: true });\n```\n\n**Important: Use file extensions when referring to relative files!**\n\n```js\nsb.mock(import('./relative/module.ts'), { spy: true });\n```\n\n2. **Specify mock values in stories**:\n   You can override the behaviour of the mocks per-story using `beforeEach` and the `mocked()` type function:\n\n```js\nimport { expect, mocked, fn } from 'storybook/test';\nimport { library } from 'some-library';\n\nconst meta = {\n  component: AuthButton,\n  beforeEach: async () => {\n    mocked(library).mockResolvedValue({  user: 'data' });\n  },\n};\n\nexport const LoggedIn: Story = {\n  play: async ({ canvas }) => {\n    await expect(library).toHaveBeenCalled();\n  },\n};\n```\n\nBefore doing this ensure you have mocked the import in the preview file.\n\n### Play Function Parameters\n\n- The play function has a `canvas` parameter that can be used directly with testing-library-like query methods.\n- It also has a `canvasElement` which is the actual DOM element.\n- The `within`-function imported from `storybook/test` transforms a DOM element to an object with query methods, similar to `canvas`.\n\n**DO NOT** use `within(canvas)` - it is redundant because `canvas` already has the query methods, `canvas` is not a DOM element.\n\n```ts\n// ✅ Correct: Use canvas directly\nplay: async ({ canvas }) => {\n	await canvas.getByLabelText('Submit').click();\n};\n\n// ⚠️ Also acceptable: Use `canvasElement` with `within`\nimport { within } from 'storybook/test';\n\nplay: async ({ canvasElement }) => {\n	const canvas = within(canvasElement);\n	await canvas.getByLabelText('Submit').click();\n};\n\n// ❌ Wrong: Do NOT use within(canvas)\nplay: async ({ canvas }) => {\n	const screen = within(canvas); // Error!\n};\n```\n\n### Key Requirements\n\n- **Node.js 20+**, **TypeScript 4.9+**\n- React Native uses `.rnstorybook` directory\n\n## Story Linking Agent Behavior\n\n- ALWAYS provide story links after any changes to stories files, including changes to existing stories.\n- {{STORY_LINKING_WORKFLOW}}\n- {{FINAL_LINKS_GUIDANCE}}\n- When you will share preview/story links (i.e. when you are not collapsing everything into a single review section), pass only the most relevant story IDs into **preview-stories** (generally no more than 5) so the returned URL list stays manageable. Include every URL returned by that call in your final response — do not drop links there.\n- {{CHANGED_STORY_FALLBACK_LINK_GUIDANCE}}\n- After changing any UI components, ALWAYS search for related stories that might cover the changes you've made. If you find any, provide the story links to the user. THIS IS VERY IMPORTANT, as it allows the user to visually inspect the changes you've made. Even later in a session when changing UI components or stories that have already been linked to previously, YOU MUST PROVIDE THE LINKS AGAIN.\n";

//#endregion
//#region src/instructions/story-testing-instructions.md
var story_testing_instructions_default = "## Story Testing Requirements\n\n**Run `{{RUN_STORY_TESTS_TOOL_NAME}}` after EVERY component or story change.** This includes creating, modifying, or refactoring components, stories, or their dependencies. It is the only way to run story tests — never run package.json test scripts (like `npm run test:stories`) instead.\n\n### Workflow\n\n1. Make your change\n2. Run `{{RUN_STORY_TESTS_TOOL_NAME}}` with related stories for focused feedback (faster while iterating)\n3. If tests fail: analyze, fix{{A11Y_FIX_SUFFIX}}, re-run\n4. Repeat until all tests pass\n\nDo not skip tests, ignore failures, or move on with failing tests. If stuck after multiple attempts, report to user.\n\n### Focused vs. full-suite test runs\n\n- Prefer focused runs (`stories` input) during development to validate the parts you changed quickly.\n- Run all tests (omit `stories`) before final handoff, after broad/refactor changes, or when impact is unclear and you need project-wide verification.\n";

//#endregion
//#region src/instructions/a11y-instructions.md
var a11y_instructions_default = "### Accessibility Violations\n\n**Fix automatically** (semantic/structural, no visual change):\n\n- ARIA attributes, roles, labels, alt text\n- Heading hierarchy, landmarks, table structure\n- Keyboard access (tabindex, focus, handlers)\n- Document-level: lang attr, frame titles, duplicate IDs\n\n**Confirm with user first** (visual/design changes):\n\n- Color contrast ratios\n- Font sizes, spacing, layout\n- Focus indicator styling\n\nDescribe the issue, ask how the user wants to proceed, and provide 2-3 concrete options.\nDo not auto-apply visual changes before user confirmation, and do not claim visual issues are fixed until they approve an option.\n";

//#endregion
//#region src/tools/get-storybook-story-instructions.ts
/**
* Injected into the story instructions when the documentation toolset is
* available. This tool output is the one channel every agent reads before
* writing UI (on both the MCP and the CLI/plugin path) and is never
* truncated, so it must carry the docs-workflow trigger — the server
* instructions only have room for a terse pointer, and agents otherwise
* default to reading library sources out of node_modules.
*/
const docsWorkflowGuidance = `

## Using library components

This Storybook exposes component documentation tools. Before creating or changing any UI, call **list-all-documentation** once to see what the design system already provides — build on existing components instead of hand-rolling duplicates — then call **get-documentation** with the \`id\` of each component you build on or get asked about, for its real props and usage examples. When multiple Storybook sources are configured, pass the \`storybookId\` from **list-all-documentation** on follow-up calls. Do this instead of reading the library's source or type definitions out of \`node_modules\` — stories show intended usage, raw types don't — and answer props/usage questions from these tools too. Never assume or invent props.`;
async function addGetUIBuildingInstructionsTool(server, enabled = () => server.ctx.custom?.toolsets?.dev ?? true, { docsAvailable = false } = {}) {
	const addonVitestAvailable = !!await getAddonVitestConstants();
	server.tool({
		name: GET_UI_BUILDING_INSTRUCTIONS_TOOL_NAME,
		title: "Storybook Story Development Instructions",
		get description() {
			const testToolsetAvailable = (server.ctx.custom?.toolsets?.test ?? true) && addonVitestAvailable;
			return getStorybookStoryInstructionsDescription({
				testToolsetAvailable,
				a11yAvailable: testToolsetAvailable && (server.ctx.custom?.a11yEnabled ?? false)
			});
		},
		enabled
	}, async () => {
		try {
			const { options, disableTelemetry } = server.ctx.custom ?? {};
			if (!options) throw new Error("Options are required in addon context");
			if (!disableTelemetry) await collectTelemetry({
				event: "tool:getUIBuildingInstructions",
				server,
				toolset: "dev"
			});
			return { content: [{
				type: "text",
				text: await buildStorybookStoryInstructions(options, {
					toolsets: server.ctx.custom?.toolsets,
					a11yEnabled: server.ctx.custom?.a11yEnabled,
					addonVitestAvailable,
					docsAvailable,
					reviewEnabled: server.ctx.custom?.reviewEnabled
				})
			}] };
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}
function getStorybookStoryInstructionsDescription({ testToolsetAvailable, a11yAvailable }) {
	return `Get comprehensive instructions for writing, testing, and fixing Storybook stories (.stories.tsx, .stories.ts, .stories.jsx, .stories.js, .stories.svelte, .stories.vue files).

CRITICAL: You MUST call this tool before:
- Creating new Storybook stories or story files
- Updating or modifying existing Storybook stories
- Adding new story variants or exports to story files
- Editing any file matching *.stories.* patterns
- Writing components that will need stories
- Editing anything that changes how the UI looks — components, styles, CSS, themes, colors, or design tokens; a shared file with no stories of its own still changes its consumers' stories${testToolsetAvailable ? `
- Running story tests or fixing test failures` : ""}${a11yAvailable ? `
- Handling accessibility (a11y) violations in stories (fix semantic issues directly; ask before visual/design changes)` : ""}

This tool provides essential Storybook-specific guidance including:
- How to structure stories correctly for Storybook 9
- Required imports (Meta, StoryObj from framework package)
- Test utility imports (from 'storybook/test')
- Story naming conventions and best practices
- Play function patterns for interactive testing
- Mocking strategies for external dependencies
- Story variants and coverage requirements${testToolsetAvailable ? `
- How to handle test failures${a11yAvailable ? " and accessibility violations" : ""}` : ""}

Even if you're familiar with Storybook, call this tool to ensure you're following the correct patterns, import paths, and conventions for this specific Storybook setup.`;
}
function getStorybookStoryInstructionsToolMetadata(options) {
	return {
		name: GET_UI_BUILDING_INSTRUCTIONS_TOOL_NAME,
		title: "Storybook Story Development Instructions",
		description: getStorybookStoryInstructionsDescription(options)
	};
}
async function buildStorybookStoryInstructions(options, { toolsets, a11yEnabled = false, addonVitestAvailable, docsAvailable = false, reviewEnabled: reviewEnabledOverride } = {}) {
	const frameworkPreset = await options.presets.apply("framework");
	const featuresPreset = await options.presets.apply("features", {});
	const changeDetectionEnabled = featuresPreset?.changeDetection ?? false;
	const reviewStatus = await getReviewStatus(options, { features: featuresPreset });
	const reviewEnabled = reviewEnabledOverride ?? reviewStatus.available;
	const framework = typeof frameworkPreset === "string" ? frameworkPreset : frameworkPreset?.name;
	const renderer = frameworkToRendererMap[framework];
	const storyLinkingWorkflow = changeDetectionEnabled ? reviewEnabled ? `After changing any component or story, call \`${GET_CHANGED_STORIES_TOOL_NAME}\` to discover the new, modified, and related stories affected by your change. Story IDs must come from that call (or a fallback discovery tool such as get-stories-by-component for shared-infrastructure changes) — never construct them from file names, export names, or memory. Feed the discovered IDs into **display-review** when the change is visually observable; use \`${PREVIEW_STORIES_TOOL_NAME}\` only while iterating on a specific story.` : `After changing UI, call \`${GET_CHANGED_STORIES_TOOL_NAME}\` first, then use \`${PREVIEW_STORIES_TOOL_NAME}\` with selected \`storyId\` values from those results.` : `After changing UI, call \`${PREVIEW_STORIES_TOOL_NAME}\` and share the most relevant links for the changes.`;
	const changedStoryFallbackLinkGuidance = changeDetectionEnabled ? `When sharing preview/story links (not when ending with a review section): if you did not pass every changed story into \`${PREVIEW_STORIES_TOOL_NAME}\`, include this Storybook fallback link so the user can view the complete changed list: \`/?statuses=affected;modified;new\`.` : `When sharing preview/story links (not when ending with a review section) and you passed only a subset into \`${PREVIEW_STORIES_TOOL_NAME}\`, mention that additional relevant stories may exist in Storybook.`;
	const docsToolsAvailable = (toolsets?.docs ?? true) && docsAvailable;
	let uiInstructions = storybook_story_instructions_default.replace("{{FRAMEWORK}}", framework).replace("{{RENDERER}}", renderer ?? framework).replace("\n{{DOCS_WORKFLOW_GUIDANCE}}", docsToolsAvailable ? docsWorkflowGuidance : "").replace("{{STORY_LINKING_WORKFLOW}}", storyLinkingWorkflow).replace("{{FINAL_LINKS_GUIDANCE}}", getFinalLinksGuidance(reviewEnabled)).replace("{{CHANGED_STORY_FALLBACK_LINK_GUIDANCE}}", changedStoryFallbackLinkGuidance);
	const resolvedAddonVitestAvailable = addonVitestAvailable ?? !!await getAddonVitestConstants();
	if ((toolsets?.test ?? true) && resolvedAddonVitestAvailable) {
		const a11yFixSuffix = a11yEnabled ? " (see a11y guidelines below)" : "";
		const storyTestingInstructions = story_testing_instructions_default.replaceAll("{{RUN_STORY_TESTS_TOOL_NAME}}", RUN_STORY_TESTS_TOOL_NAME).replace("{{A11Y_FIX_SUFFIX}}", a11yFixSuffix);
		uiInstructions += `\n\n${storyTestingInstructions}`;
		if (a11yEnabled) uiInstructions += `\n${a11y_instructions_default}`;
	}
	return uiInstructions;
}
const frameworkToRendererMap = {
	"@storybook/react-vite": "@storybook/react",
	"@storybook/react-webpack5": "@storybook/react",
	"@storybook/nextjs": "@storybook/react",
	"@storybook/nextjs-vite": "@storybook/react",
	"@storybook/react-native-web-vite": "@storybook/react",
	"@storybook/vue3-vite": "@storybook/vue3",
	"@nuxtjs/storybook": "@storybook/vue3",
	"@storybook/angular": "@storybook/angular",
	"@storybook/svelte-vite": "@storybook/svelte",
	"@storybook/sveltekit": "@storybook/svelte",
	"@storybook/preact-vite": "@storybook/preact",
	"@storybook/web-components-vite": "@storybook/web-components",
	"@storybook/html-vite": "@storybook/html"
};

//#endregion
//#region src/utils/build-args-param.ts
const HEX_REGEXP = /^#([a-f0-9]{3,4}|[a-f0-9]{6}|[a-f0-9]{8})$/i;
const COLOR_REGEXP = /^(rgba?|hsla?)\(([0-9]{1,3}),\s?([0-9]{1,3})%?,\s?([0-9]{1,3})%?,?\s?([0-9](\.[0-9]{1,2})?)?\)$/i;
/**
* Encodes special values for Storybook's args URL format.
* Handles undefined, null, booleans, dates, hex colors, and rgba/hsla colors.
*/
function encodeSpecialValues(value) {
	if (value === void 0) return "!undefined";
	if (value === null) return "!null";
	if (typeof value === "string") {
		if (HEX_REGEXP.test(value)) return `!hex(${value.slice(1)})`;
		if (COLOR_REGEXP.test(value)) return `!${value.replace(/[\s%]/g, "")}`;
		return value;
	}
	if (typeof value === "boolean") return `!${value}`;
	if (value instanceof Date) return `!date(${value.toISOString()})`;
	if (Array.isArray(value)) return value.map(encodeSpecialValues);
	if (typeof value === "object" && value !== null) return Object.entries(value).reduce((acc, [key, val]) => Object.assign(acc, { [key]: encodeSpecialValues(val) }), {});
	return value;
}
/**
* Replaces some url-encoded characters with their decoded equivalents.
* The URI RFC specifies these should be encoded, but all browsers will
* tolerate them being decoded, so we opt to go with it for cleaner looking URIs.
*/
function decodeKnownQueryChar(chr) {
	switch (chr) {
		case "%20": return "+";
		case "%5B": return "[";
		case "%5D": return "]";
		case "%2C": return ",";
		case "%3A": return ":";
	}
	return chr;
}
const KNOWN_QUERY_CHAR_REGEXP = /%[0-9A-F]{2}/g;
/**
* Builds a Storybook args query parameter string from an object of props.
*
* The format uses semicolons as delimiters and colons for key:value pairs,
* with special encoding for booleans, null, undefined, dates, and colors.
*
* Example output: "disabled:!true;label:Hello+World;count:42"
*/
function buildArgsParam(args) {
	if (!args || Object.keys(args).length === 0) return "";
	return stringify(encodeSpecialValues(args), {
		delimiter: ";",
		nesting: true,
		nestingSyntax: "js"
	}).replace(KNOWN_QUERY_CHAR_REGEXP, decodeKnownQueryChar).split(";").map((part) => part.replace("=", ":")).join(";");
}

//#endregion
//#region src/tools/preview-stories/preview-stories-app-template.html
var preview_stories_app_template_default = "<!doctype html>\n<html>\n	<head>\n		<meta charset=\"utf-8\" />\n		<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\" />\n		<style>\n			* {\n				margin: 0;\n				padding: 0;\n				box-sizing: border-box;\n			}\n			html[data-theme='light'] {\n				color-scheme: light;\n			}\n			html[data-theme='dark'] {\n				color-scheme: dark;\n			}\n			html,\n			body {\n				width: 100%;\n				margin: 0;\n				padding: 0;\n				background-color: var(--color-background-secondary);\n			}\n			body {\n				display: flex;\n				flex-direction: column;\n				gap: 1rem;\n			}\n			:root {\n				/*\n				These are fallback values, if the MCP client doesn't set them.\n				See https://github.com/modelcontextprotocol/ext-apps/blob/main/specification/draft/apps.mdx#theming\n				*/\n				--color-text-primary: light-dark(black, white);\n				--color-border-primary: light-dark(#ccc, #444);\n				--color-background-secondary: light-dark(#f9f9f9, #1e1e1e);\n				--font-sans:\n					-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Oxygen-Sans', Ubuntu, Cantarell,\n					'Helvetica Neue', sans-serif;\n				--font-heading-xs-size: 1rem;\n				--font-heading-xs-line-height: 1.25;\n				--border-width-regular: 1px;\n			}\n			.story-heading {\n				font-family: var(--font-sans);\n				font-size: var(--font-heading-xs-size);\n				line-height: var(--font-heading-xs-line-height);\n				color: var(--color-text-primary);\n				padding: 0.5rem 0;\n			}\n			.story-iframe {\n				background-color: white;\n				border: var(--border-width-regular) solid var(--color-border-primary);\n			}\n			body:has(> article:only-child) h1 {\n				/* if there is only one story rendered, hide its heading */\n				display: none;\n			}\n		</style>\n		<template id=\"preview-template\">\n			<article>\n				<h1 class=\"story-heading\"></h1>\n				<iframe class=\"story-iframe\"></iframe>\n			</article>\n		</template>\n		<script type=\"module\">\n			// APP_SCRIPT_PLACEHOLDER\n		<\/script>\n	</head>\n	<body></body>\n</html>\n";

//#endregion
//#region src/tools/preview-stories.ts
const PREVIEW_STORIES_RESOURCE_URI = `ui://${PREVIEW_STORIES_TOOL_NAME}/preview.html`;
function getPreviewStoriesToolDescription({ reviewEnabled = false } = {}) {
	if (!reviewEnabled) return `Use this tool to get one or more Storybook preview URLs.
Call it after editing anything that changes how the UI looks — components, stories, styles, CSS, themes, colors, or design tokens — no exceptions. A shared file has no stories of its own: preview the stories of the components that consume it.
Include each returned preview URL in your final user-facing response so users can open them directly.`;
	return `Use this tool to get Storybook preview URLs while iterating on a specific story, or when the user asks for a direct link to one.
Do not end visual work or browse requests with these links — publish a curated review with display-review instead (passing changedFiles: [] when no code changed) and link that.`;
}
const PreviewStoriesInput = v.object({ stories: v.pipe(StoryInputArray, v.description(`Stories to preview.
Prefer { storyId } when you don't already have story file context, since this avoids filesystem discovery.
Use { storyId } when IDs were discovered from documentation tools.
Use { absoluteStoryPath + exportName } only when you're already working in a specific .stories.* file and already have that context.`)) });
const PreviewStoriesOutput = v.object({ stories: v.array(v.union([v.object({
	title: v.string(),
	name: v.string(),
	previewUrl: v.pipe(v.string(), v.description("Direct URL to open the story preview. Include this URL in the final user-facing response so users can open it directly — unless a curated review page is being published via display-review, in which case link the review page instead of listing individual URLs."))
}), v.object({
	input: StoryInput,
	error: v.string()
})])) });
function getPreviewStoriesToolMetadata({ reviewEnabled = false } = {}) {
	return {
		name: PREVIEW_STORIES_TOOL_NAME,
		title: "Get story preview URLs",
		description: getPreviewStoriesToolDescription({ reviewEnabled }),
		schema: PreviewStoriesInput,
		outputSchema: PreviewStoriesOutput,
		_meta: { ui: { resourceUri: PREVIEW_STORIES_RESOURCE_URI } }
	};
}
async function addPreviewStoriesTool(server, enabled = () => server.ctx.custom?.toolsets?.dev ?? true, { reviewEnabled = false } = {}) {
	const previewStoryAppScript = await fs$1.readFile(url.fileURLToPath(import.meta.resolve("@storybook/addon-mcp/internal/preview-stories-app-script")), "utf-8");
	const appHtml = preview_stories_app_template_default.replace("// APP_SCRIPT_PLACEHOLDER", previewStoryAppScript);
	server.resource({
		name: PREVIEW_STORIES_RESOURCE_URI,
		description: "App resource for the Preview Stories tool",
		uri: PREVIEW_STORIES_RESOURCE_URI,
		mimeType: "text/html;profile=mcp-app"
	}, () => {
		const origin = server.ctx.custom.origin;
		return { contents: [{
			uri: PREVIEW_STORIES_RESOURCE_URI,
			mimeType: "text/html;profile=mcp-app",
			text: appHtml,
			_meta: { ui: {
				prefersBorder: false,
				domain: origin,
				csp: {
					connectDomains: [origin],
					resourceDomains: [origin],
					frameDomains: [origin],
					baseUriDomains: [origin]
				}
			} }
		}] };
	});
	server.tool({
		...getPreviewStoriesToolMetadata({ reviewEnabled }),
		enabled
	}, async (input) => {
		try {
			const { origin, options, disableTelemetry } = server.ctx.custom ?? {};
			if (!origin) throw new Error("Origin is required in addon context");
			if (!options) throw new Error("Storybook options are required in addon context");
			const index = await getStoryIndex(options);
			const resolvedStories = findStoryIds(index, input.stories);
			const structuredResult = [];
			const textResult = [];
			for (const story of resolvedStories) {
				if ("errorMessage" in story) {
					structuredResult.push({
						input: story.input,
						error: story.errorMessage
					});
					textResult.push(story.errorMessage);
					continue;
				}
				const indexEntry = index.entries[story.id];
				if (!indexEntry) {
					structuredResult.push({
						input: story.input,
						error: `No story found for story ID "${story.id}"`
					});
					textResult.push(`No story found for story ID "${story.id}"`);
					continue;
				}
				let previewUrl = `${origin}/?path=/story/${story.id}`;
				const argsParam = buildArgsParam(story.input.props ?? {});
				if (argsParam) previewUrl += `&args=${argsParam}`;
				const globalsParam = buildArgsParam(story.input.globals ?? {});
				if (globalsParam) previewUrl += `&globals=${globalsParam}`;
				structuredResult.push({
					title: indexEntry.title,
					name: indexEntry.name,
					previewUrl
				});
				textResult.push(previewUrl);
			}
			if (!disableTelemetry) await collectTelemetry({
				event: "tool:previewStories",
				server,
				toolset: "dev",
				inputStoryCount: input.stories.length,
				outputStoryCount: structuredResult.length
			});
			const content = textResult.map((text) => ({
				type: "text",
				text
			}));
			const resolvedAnyStory = structuredResult.some((story) => "previewUrl" in story);
			if (reviewEnabled && resolvedAnyStory) content.push({
				type: "text",
				text: `These preview links are for iterating or sharing a specific story — they are not how visual work or a browse request ends. The display-review tool is available in this session: if you are finishing visually observable work or showing a set of stories, publish the review with **display-review** and link that instead.`
			});
			return {
				content,
				structuredContent: { stories: structuredResult }
			};
		} catch (error) {
			return errorToMCPContent(error);
		}
	});
}

//#endregion
//#region src/tools/tool-registry.ts
const isToolsetEnabled = (toolset, toolsets) => toolsets?.[toolset] ?? true;
const isToolAvailable = (definition, context) => definition.available?.(context) ?? true;
const isMetadataToolEnabled = (definition, context) => isToolsetEnabled(definition.toolset, context.toolsets) && isToolAvailable(definition, context);
const createToolsetEnabled = (server, toolset) => () => server.ctx.custom?.toolsets?.[toolset] ?? true;
const addonToolDefinitions = [
	{
		name: PREVIEW_STORIES_TOOL_NAME,
		toolset: "dev",
		getMetadata: ({ availability }) => getPreviewStoriesToolMetadata({ reviewEnabled: availability.reviewEnabled }),
		register: (server, { availability }, enabled) => addPreviewStoriesTool(server, enabled, { reviewEnabled: availability.reviewEnabled })
	},
	{
		name: GET_UI_BUILDING_INSTRUCTIONS_TOOL_NAME,
		toolset: "dev",
		getMetadata: ({ availability, toolsets }) => {
			const testToolsetAvailable = isToolsetEnabled("test", toolsets) && availability.testSupported;
			return getStorybookStoryInstructionsToolMetadata({
				testToolsetAvailable,
				a11yAvailable: testToolsetAvailable && availability.a11yEnabled
			});
		},
		register: (server, { availability, toolsets }, enabled) => addGetUIBuildingInstructionsTool(server, enabled, { docsAvailable: isToolsetEnabled("docs", toolsets) && availability.docsEnabled }),
		getLocalTool: ({ availability, toolsets, options }) => ({ call: async () => {
			return { content: [{
				type: "text",
				text: await buildStorybookStoryInstructions(options, {
					toolsets,
					a11yEnabled: availability.a11yEnabled,
					addonVitestAvailable: availability.testSupported,
					docsAvailable: isToolsetEnabled("docs", toolsets) && availability.docsEnabled,
					reviewEnabled: availability.reviewEnabled
				})
			}] };
		} })
	},
	{
		name: GET_CHANGED_STORIES_TOOL_NAME,
		toolset: "dev",
		available: ({ availability }) => availability.changeDetectionEnabled,
		getMetadata: () => getChangedStoriesToolMetadata(),
		register: (server, { availability }, enabled) => addGetChangedStoriesTool(server, enabled, { reviewEnabled: availability.reviewEnabled })
	},
	{
		name: GET_STORIES_BY_COMPONENT_TOOL_NAME,
		toolset: "dev",
		available: ({ availability }) => availability.moduleGraphSupported,
		getMetadata: ({ availability }) => getStoriesByComponentToolMetadata({ reviewEnabled: availability.reviewEnabled }),
		register: (server, { availability }, enabled) => addGetStoriesByComponentTool(server, enabled, { reviewEnabled: availability.reviewEnabled })
	},
	{
		name: DISPLAY_REVIEW_TOOL_NAME,
		toolset: "dev",
		available: ({ availability }) => availability.reviewEnabledForCli,
		getMetadata: () => getDisplayReviewToolMetadata(),
		register: (server, { availability }, enabled) => addDisplayReviewTool(server, async () => (await enabled?.() ?? true) && (server.ctx.custom?.reviewEnabled ?? availability.reviewEnabled))
	},
	{
		name: RUN_STORY_TESTS_TOOL_NAME,
		toolset: "test",
		available: ({ availability }) => availability.testSupported,
		getMetadata: ({ availability }) => getRunStoryTestsToolMetadata({ a11yEnabled: availability.a11yEnabled }),
		register: (server, { availability }, enabled) => addRunStoryTestsTool(server, { a11yEnabled: availability.a11yEnabled }, enabled)
	},
	{
		name: LIST_TOOL_NAME,
		toolset: "docs",
		available: ({ availability }) => availability.docsEnabled,
		getMetadata: () => getListAllDocumentationToolMetadata(),
		register: async (server, _context, enabled) => {
			logger.info("Experimental components manifest feature detected - registering component tools");
			await addListAllDocumentationTool(server, enabled);
		}
	},
	{
		name: GET_TOOL_NAME,
		toolset: "docs",
		available: ({ availability }) => availability.docsEnabled,
		getMetadata: ({ multiSource }) => getDocumentationToolMetadata({ multiSource }),
		register: (server, { multiSource }, enabled) => addGetDocumentationTool(server, enabled, { multiSource })
	},
	{
		name: GET_STORY_TOOL_NAME,
		toolset: "docs",
		available: ({ availability }) => availability.docsEnabled,
		getMetadata: ({ multiSource }) => getStoryDocumentationToolMetadata({ multiSource }),
		register: (server, { multiSource }, enabled) => addGetStoryDocumentationTool(server, enabled, { multiSource })
	}
];
function getAddonToolMetadata(context) {
	return addonToolDefinitions.filter((definition) => isMetadataToolEnabled(definition, context)).map((definition) => definition.getMetadata(context));
}
function getAddonLocalTools(context) {
	return Object.fromEntries(addonToolDefinitions.filter((definition) => isMetadataToolEnabled(definition, context)).flatMap((definition) => {
		const localTool = definition.getLocalTool?.(context);
		return localTool ? [[definition.name, localTool]] : [];
	}));
}
async function registerAddonMcpTools(server, context) {
	for (const definition of addonToolDefinitions) if (isToolsetEnabled(definition.toolset, context.toolsets) && isToolAvailable(definition, context)) await definition.register(server, context, createToolsetEnabled(server, definition.toolset));
}

//#endregion
//#region src/mcp-handler.ts
let transport;
let origin;
let initialize;
let disableTelemetry;
let a11yEnabled;
let reviewGates;
const initializeMCPServer = async (options, multiSource) => {
	const core = await options.presets.apply("core", {});
	const features = await options.presets.apply("features", {});
	disableTelemetry = core?.disableTelemetry ?? false;
	const availability = getEffectiveToolAvailability(await getToolAvailability(options, { features }), { multiSource });
	a11yEnabled = availability.a11yEnabled;
	reviewGates = {
		reviewEnabled: availability.reviewEnabled,
		reviewEnabledForCli: availability.reviewEnabledForCli
	};
	let server;
	const serverOptions = {
		adapter: new ValibotJsonSchemaAdapter(),
		get instructions() {
			return buildServerInstructions({
				devEnabled: server?.ctx.custom?.toolsets?.dev ?? true,
				testEnabled: (server?.ctx.custom?.toolsets?.test ?? true) && availability.testSupported,
				docsEnabled: (server?.ctx.custom?.toolsets?.docs ?? true) && availability.docsEnabled,
				changeDetectionEnabled: availability.changeDetectionEnabled,
				moduleGraphSupported: availability.moduleGraphSupported,
				reviewEnabled: server?.ctx.custom?.reviewEnabled ?? availability.reviewEnabled
			});
		},
		capabilities: {
			tools: { listChanged: true },
			resources: { listChanged: true }
		}
	};
	server = new McpServer({
		name,
		version,
		description
	}, serverOptions).withContext();
	if (!disableTelemetry) server.on("initialize", async () => {
		await collectTelemetry({
			event: "session:initialized",
			server
		});
	});
	await registerAddonMcpTools(server, {
		availability,
		multiSource
	});
	transport = new HttpTransport(server, { path: null });
	origin = `http://localhost:${options.port}`;
	logger.debug(`MCP server origin: ${origin}`);
	return server;
};
const mcpServerHandler = async ({ req, res, options, addonOptions, endpoint = DEFAULT_MCP_ENDPOINT, sources, manifestProvider, resolveEntry, compositionAuth }) => {
	if (!initialize) initialize = initializeMCPServer(options, sources?.some((s) => s.url));
	const server = await initialize;
	const webRequest = await incomingMessageToWebRequest(req);
	const addonContext = {
		options,
		endpoint,
		toolsets: getToolsets(webRequest, addonOptions),
		reviewEnabled: isReviewEnabledForRequest(webRequest, reviewGates),
		origin,
		disableTelemetry,
		a11yEnabled,
		request: webRequest,
		sources,
		manifestProvider,
		resolveEntry,
		...!disableTelemetry && {
			onListAllDocumentation: async ({ manifests, resultText, sources: sourceManifests }) => {
				await collectTelemetry({
					event: "tool:listAllDocumentation",
					server,
					toolset: "docs",
					componentCount: Object.keys(manifests.componentManifest.components).length,
					docsCount: Object.keys(manifests.docsManifest?.docs || {}).length,
					resultTokenCount: estimateTokens(resultText),
					sourceCount: sourceManifests?.length
				});
			},
			onGetDocumentation: async ({ input, foundDocumentation, resultText }) => {
				await collectTelemetry({
					event: "tool:getDocumentation",
					server,
					toolset: "docs",
					componentId: input.id,
					found: !!foundDocumentation,
					resultTokenCount: estimateTokens(resultText ?? "")
				});
			}
		}
	};
	const response = await transport.respond(webRequest, addonContext);
	if (response) {
		const body = await response.arrayBuffer();
		await webResponseToServerResponse(compositionAuth.hadAuthError(webRequest) ? new Response("401 - Unauthorized", {
			status: 401,
			headers: {
				"Content-Type": "text/plain",
				"WWW-Authenticate": compositionAuth.buildWwwAuthenticate(origin)
			}
		}) : new Response(body, {
			status: response.status,
			headers: response.headers
		}), res);
	}
};
/**
* Converts a Node.js IncomingMessage to a Web Request.
*/
async function incomingMessageToWebRequest(req) {
	const host = req.headers.host || "localhost";
	const protocol = "encrypted" in req.socket && req.socket.encrypted ? "https" : "http";
	const url = new URL(req.url || "/", `${protocol}://${host}`);
	const bodyBuffer = await buffer(req);
	return new Request(url, {
		method: req.method,
		headers: req.headers,
		body: bodyBuffer.length > 0 ? new Uint8Array(bodyBuffer) : void 0
	});
}
/**
* Converts a Web Response to a Node.js ServerResponse.
*/
async function webResponseToServerResponse(webResponse, nodeResponse) {
	nodeResponse.statusCode = webResponse.status;
	webResponse.headers.forEach((value, key) => {
		nodeResponse.setHeader(key, value);
	});
	if (webResponse.body) {
		const reader = webResponse.body.getReader();
		try {
			while (true) {
				const { done, value } = await reader.read();
				if (done) break;
				nodeResponse.write(value);
			}
		} finally {
			reader.releaseLock();
		}
	}
	nodeResponse.end();
}
/**
* Review is on for a request when the `experimentalReview` flag enables it
* globally, or when the request marks itself as coming from the `storybook ai`
* CLI (the Claude/Codex plugins) via {@link STORYBOOK_MCP_PROXY_HEADER} —
* that channel gets review by default, direct MCP clients stay opt-in.
*/
function isReviewEnabledForRequest(request, gates) {
	return gates.reviewEnabled || gates.reviewEnabledForCli && request.headers.get(STORYBOOK_MCP_PROXY_HEADER) === "true";
}
function getToolsets(request, addonOptions) {
	const toolsetHeader = request.headers.get("X-MCP-Toolsets");
	if (!toolsetHeader || toolsetHeader.trim() === "") return addonOptions.toolsets;
	const toolsets = {
		dev: false,
		docs: false,
		test: false
	};
	const enabledToolsets = toolsetHeader.split(",");
	for (const enabledToolset of enabledToolsets) {
		const trimmedToolset = enabledToolset.trim();
		if (trimmedToolset in toolsets) toolsets[trimmedToolset] = true;
	}
	return toolsets;
}

//#endregion
//#region src/template.html
var template_default = "<!doctype html>\n<html>\n	<head>\n		<style>\n			@font-face {\n				font-family: 'Nunito Sans';\n				font-style: normal;\n				font-weight: 400;\n				font-display: swap;\n				src: url('./sb-common-assets/nunito-sans-regular.woff2') format('woff2');\n			}\n\n			* {\n				margin: 0;\n				padding: 0;\n				box-sizing: border-box;\n			}\n\n			html,\n			body {\n				height: 100%;\n				font-family:\n					'Nunito Sans',\n					-apple-system,\n					BlinkMacSystemFont,\n					'Segoe UI',\n					Roboto,\n					Oxygen,\n					Ubuntu,\n					Cantarell,\n					sans-serif;\n			}\n\n			body {\n				display: flex;\n				flex-direction: column;\n				justify-content: center;\n				align-items: center;\n				text-align: center;\n				padding: 2rem;\n				background-color: #ffffff;\n				color: rgb(46, 52, 56);\n				line-height: 1.6;\n			}\n\n			p {\n				margin-bottom: 1rem;\n			}\n\n			code {\n				font-family: 'Monaco', 'Courier New', monospace;\n				background: #f5f5f5;\n				padding: 0.2em 0.4em;\n				border-radius: 3px;\n			}\n\n			a {\n				color: #1ea7fd;\n			}\n\n			.container {\n				display: flex;\n				flex-direction: column;\n				align-items: center;\n			}\n\n			.toolsets {\n				margin: 1.5rem 0;\n				text-align: left;\n				max-width: 500px;\n			}\n\n			.toolsets h3 {\n				font-size: 1rem;\n				margin-bottom: 0.75rem;\n				text-align: center;\n			}\n\n			.toolset {\n				margin-bottom: 1rem;\n				padding: 0.75rem 1rem;\n				border-radius: 6px;\n				background: #f8f9fa;\n				border: 1px solid #e9ecef;\n			}\n\n			.toolset-header {\n				display: flex;\n				align-items: center;\n				gap: 0.5rem;\n				font-weight: 600;\n				margin-bottom: 0.5rem;\n			}\n\n			.toolset-status {\n				display: inline-block;\n				padding: 0.15em 0.5em;\n				border-radius: 3px;\n				font-size: 0.75rem;\n				font-weight: 500;\n				text-transform: uppercase;\n			}\n\n			.toolset-status.enabled {\n				background: #d4edda;\n				color: #155724;\n			}\n\n			.toolset-status.disabled {\n				background: #f8d7da;\n				color: #721c24;\n			}\n\n			.toolset-tools {\n				font-size: 0.875rem;\n				color: #6c757d;\n				padding-left: 1.5rem;\n				margin: 0;\n			}\n\n			.toolset-tools li {\n				margin-bottom: 0.25rem;\n			}\n\n			.toolset-tools code {\n				font-size: 0.8rem;\n			}\n\n			.toolset-notice {\n				font-size: 0.8rem;\n				color: #856404;\n				background: #fff3cd;\n				padding: 0.5rem;\n				border-radius: 4px;\n				margin-top: 0.5rem;\n			}\n\n			.toolset-notice a {\n				color: #533f03;\n			}\n\n			@media (prefers-color-scheme: dark) {\n				body {\n					background-color: rgb(34, 36, 37);\n					color: rgb(201, 205, 207);\n				}\n\n				code {\n					background: rgba(255, 255, 255, 0.1);\n				}\n\n				.toolset {\n					background: rgba(255, 255, 255, 0.05);\n					border-color: rgba(255, 255, 255, 0.1);\n				}\n\n				.toolset-tools {\n					color: #adb5bd;\n				}\n\n				.toolset-status.enabled {\n					background: rgba(40, 167, 69, 0.2);\n					color: #75d67e;\n				}\n\n				.toolset-status.disabled {\n					background: rgba(220, 53, 69, 0.2);\n					color: #f5a6ad;\n				}\n\n				.toolset-notice {\n					background: rgba(255, 193, 7, 0.15);\n					color: #ffc107;\n				}\n\n				.toolset-notice a {\n					color: #ffe066;\n				}\n			}\n		</style>\n	</head>\n	<body>\n		<div class=\"container\">\n			<p>\n				Storybook MCP server successfully running via\n				<code>@storybook/addon-mcp</code>.\n			</p>\n			<p>\n				See how to connect to it from your coding agent in\n				<a\n					target=\"_blank\"\n					href=\"https://github.com/storybookjs/mcp/tree/main/packages/addon-mcp#configuring-your-agent\"\n					>the addon's README</a\n				>.\n			</p>\n\n			<div class=\"toolsets\">\n				<h3>Available Toolsets</h3>\n\n				<div class=\"toolset\">\n					<div class=\"toolset-header\">\n						<span>dev</span>\n						<span class=\"toolset-status {{DEV_STATUS}}\">{{DEV_STATUS}}</span>\n					</div>\n					<ul class=\"toolset-tools\">\n						<li><code>preview-stories</code></li>\n						<li><code>get-storybook-story-instructions</code></li>\n						<li>\n							<code>get-changed-stories</code>\n							<span class=\"toolset-status {{CHANGE_DETECTION_STATUS}}\"\n								>{{CHANGE_DETECTION_STATUS}}</span\n							>\n						</li>\n						<li>\n							<code>get-stories-by-component</code>\n							<span class=\"toolset-status {{STORIES_BY_COMPONENT_STATUS}}\"\n								>{{STORIES_BY_COMPONENT_STATUS}}</span\n							>\n						</li>\n						<li>\n							<code>display-review</code>\n							<span class=\"toolset-status {{REVIEW_STATUS}}\">{{REVIEW_STATUS}}</span>\n						</li>\n					</ul>\n					{{DEV_NOTICE}}\n				</div>\n\n				<div class=\"toolset\">\n					<div class=\"toolset-header\">\n						<span>docs</span>\n						<span class=\"toolset-status {{DOCS_STATUS}}\">{{DOCS_STATUS}}</span>\n					</div>\n					<ul class=\"toolset-tools\">\n						<li><code>list-all-documentation</code></li>\n						<li><code>get-documentation</code></li>\n						<li><code>get-documentation-for-story</code></li>\n					</ul>\n					{{DOCS_NOTICE}}\n				</div>\n\n				<div class=\"toolset\">\n					<div class=\"toolset-header\">\n						<span>test</span>\n						<span class=\"toolset-status {{TEST_STATUS}}\">{{TEST_STATUS}}</span>\n					</div>\n					<ul class=\"toolset-tools\">\n						<li><code>run-story-tests</code>{{A11Y_BADGE}}</li>\n					</ul>\n					{{TEST_NOTICE}}\n				</div>\n			</div>\n\n			{{MANIFEST_DEBUGGER_LINK}}\n		</div>\n	</body>\n</html>\n";

//#endregion
//#region src/auth/composition-auth.ts
/**
* Composition authentication for fetching manifests from private Storybooks.
*
* This class handles OAuth discovery and token-based manifest fetching for
* composed Storybooks (refs). It acts as a proxy for OAuth metadata, allowing
* MCP clients like VS Code to handle the OAuth flow with Chromatic.
*/
const OAuthResourceMetadata = v.object({
	resource: v.optional(v.string()),
	authorization_servers: v.pipe(v.array(v.string()), v.minLength(1)),
	scopes_supported: v.optional(v.array(v.string()))
});
const OAuthServerMetadata = v.object({
	issuer: v.string(),
	authorization_endpoint: v.string(),
	token_endpoint: v.string(),
	scopes_supported: v.optional(v.array(v.string()))
});
const MANIFEST_CACHE_TTL = 3600 * 1e3;
const REVALIDATION_TTL = 60 * 1e3;
var AuthenticationError = class extends Error {
	constructor(url) {
		super(`Authentication failed for ${url}. Your token may be invalid or expired.`);
		this.name = "AuthenticationError";
	}
};
var CompositionAuth = class {
	#authRequirement = null;
	#authRequiredUrls = [];
	#refsWithManifests = [];
	#manifestCache = /* @__PURE__ */ new Map();
	#lastToken = null;
	#authErrors = /* @__PURE__ */ new WeakMap();
	/** Initialize by checking which refs require authentication and have manifests. */
	async initialize(refs) {
		for (const ref of refs) try {
			const result = await this.#checkRef(ref.url);
			if (result === "no-manifest") continue;
			this.#refsWithManifests.push(ref);
			if (result === "public") continue;
			this.#authRequiredUrls.push(ref.url);
			if (!this.#authRequirement) this.#authRequirement = result;
			else {
				const existingServer = this.#authRequirement.resourceMetadata.authorization_servers[0];
				const newServer = result.resourceMetadata.authorization_servers[0];
				if (existingServer !== newServer) console.warn(`[addon-mcp] Composed ref "${ref.title}" uses a different OAuth server (${newServer}) than the first authenticated ref (${existingServer}). Only the first OAuth server will be used for authentication.`);
			}
		} catch (error) {
			console.warn(`[addon-mcp] Failed to check auth for composed ref "${ref.title}" (${ref.url}): ${error instanceof Error ? error.message : String(error)}. Skipping this ref.`);
		}
	}
	get requiresAuth() {
		return this.#authRequiredUrls.length > 0;
	}
	get authUrls() {
		return this.#authRequiredUrls;
	}
	/** Check if a request encountered an auth error during manifest fetching. */
	hadAuthError(request) {
		return this.#authErrors.has(request);
	}
	/** Check if a URL requires authentication based on discovered auth requirements. */
	#isAuthRequiredUrl(url) {
		return this.#authRequiredUrls.some((authUrl) => url.startsWith(authUrl));
	}
	/** Build .well-known/oauth-protected-resource response. */
	buildWellKnown(origin) {
		if (!this.#authRequirement) return null;
		return {
			resource: `${origin}/mcp`,
			authorization_servers: this.#authRequirement.resourceMetadata.authorization_servers,
			scopes_supported: this.#authRequirement.resourceMetadata.scopes_supported
		};
	}
	/** Build WWW-Authenticate header for 401 responses */
	buildWwwAuthenticate(origin) {
		return `Bearer error="unauthorized", error_description="Authorization needed for composed Storybooks", resource_metadata="${origin}/.well-known/oauth-protected-resource"`;
	}
	/** Build sources configuration: local first, then refs that have manifests. */
	buildSources() {
		return [{
			id: "local",
			title: "Local"
		}, ...this.#refsWithManifests.map((ref) => ({
			id: ref.id,
			title: ref.title,
			url: ref.url
		}))];
	}
	/**
	* Create a manifest provider for multi-source mode.
	*
	* Remote sources are always fetched over HTTP (with auth + caching). The local
	* source normally fetches from `localOrigin`, but in `experimentalDocgenServer`
	* mode core 404s `/manifests/*.json` (the data lives in the open services), so
	* callers pass `localManifestProvider` to read the local source in-process
	* instead. When omitted, the local source keeps its HTTP behavior.
	*/
	createManifestProvider(localOrigin, localManifestProvider) {
		return async (request, path, source) => {
			const token = extractBearerToken(request?.headers.get("Authorization"));
			const remoteSource = source?.url ? {
				...source,
				url: source.url
			} : void 0;
			if (!remoteSource && localManifestProvider) return localManifestProvider(request, path, source);
			const baseUrl = remoteSource?.url ?? localOrigin;
			const manifestUrl = `${baseUrl}${path.replace("./", "/")}`;
			const isRemote = !!remoteSource;
			const tokenForRequest = isRemote && this.#isAuthRequiredUrl(baseUrl) ? token : null;
			if (token && token !== this.#lastToken) {
				this.#manifestCache.clear();
				this.#lastToken = token;
			}
			if (isRemote) {
				const cached = this.#manifestCache.get(manifestUrl);
				if (cached) if (Date.now() - cached.timestamp > MANIFEST_CACHE_TTL) this.#manifestCache.delete(manifestUrl);
				else {
					if (!cached.lastRevalidatedAt || Date.now() - cached.lastRevalidatedAt > REVALIDATION_TTL) {
						cached.lastRevalidatedAt = Date.now();
						this.#fetchManifest(manifestUrl, tokenForRequest).then((text) => this.#manifestCache.set(manifestUrl, {
							text,
							timestamp: Date.now(),
							lastRevalidatedAt: Date.now()
						})).catch(() => {});
					}
					return cached.text;
				}
			}
			try {
				const text = await this.#fetchManifest(manifestUrl, tokenForRequest);
				if (isRemote) this.#manifestCache.set(manifestUrl, {
					text,
					timestamp: Date.now()
				});
				return text;
			} catch (error) {
				if (error instanceof AuthenticationError && request) this.#authErrors.set(request, error);
				throw error;
			}
		};
	}
	/**
	* Fetch a manifest with optional auth token.
	* If the response is 200 but not a valid manifest, checks /mcp for auth issues.
	*/
	async #fetchManifest(url, token) {
		const headers = { Accept: "application/json" };
		if (token) headers["Authorization"] = `Bearer ${token}`;
		const response = await fetch(url, { headers });
		if (response.status === 401) throw new AuthenticationError(url);
		if (!response.ok) throw new Error(`Failed to fetch ${url}: ${response.status}`);
		const text = await response.text();
		const { pathname } = new URL(url);
		const isComponentsManifest = pathname.endsWith("/manifests/components.json");
		const isDocsManifest = pathname.endsWith("/manifests/docs.json");
		const schema = isComponentsManifest ? ComponentManifestMap : isDocsManifest ? DocsManifestMap : v.unknown();
		if (v.safeParse(v.pipe(v.string(), v.parseJson(), schema), text).success) return text;
		if (await this.#isMcpUnauthorized(new URL(url).origin)) throw new AuthenticationError(url);
		throw new Error(`Invalid manifest response from ${url}: expected valid JSON manifest but got unexpected content.`);
	}
	/**
	* Check a ref to determine if it has a manifest and whether it requires auth.
	* Returns 'public' if the ref has a valid manifest without auth,
	* 'no-manifest' if no manifest is available, or an AuthRequirement if auth is needed.
	*/
	async #checkRef(refUrl) {
		const response = await fetch(`${refUrl}/manifests/components.json`, { headers: { Accept: "application/json" } });
		const authReq = await this.#parseAuthFromResponse(response);
		if (authReq) return authReq;
		if (response.ok) {
			const text = await response.text();
			if (v.safeParse(v.pipe(v.string(), v.parseJson(), ComponentManifestMap), text).success) return "public";
		}
		const mcpAuth = await this.#checkMcpAuth(refUrl);
		if (mcpAuth) return mcpAuth;
		return "no-manifest";
	}
	/** Check /mcp endpoint for 401 auth requirement. */
	async #checkMcpAuth(refUrl) {
		const response = await fetch(`${refUrl}/mcp`, {
			method: "POST",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({
				jsonrpc: "2.0",
				id: 1,
				method: "tools/list"
			})
		});
		return this.#parseAuthFromResponse(response);
	}
	/** Quick check: does the remote /mcp return 401? */
	async #isMcpUnauthorized(origin) {
		try {
			return (await fetch(`${origin}/mcp`, {
				method: "POST",
				headers: { "Content-Type": "application/json" },
				body: JSON.stringify({
					jsonrpc: "2.0",
					id: 1,
					method: "tools/list"
				})
			})).status === 401;
		} catch {
			return false;
		}
	}
	/** Extract auth requirement from a 401 response's WWW-Authenticate header. */
	async #parseAuthFromResponse(response) {
		if (response.status !== 401) return null;
		const wwwAuth = response.headers.get("WWW-Authenticate");
		if (!wwwAuth) return null;
		const match = wwwAuth.match(/resource_metadata="([^"]+)"/);
		if (!match?.[1]) return null;
		const resourceMetadataUrl = match[1];
		const resourceResponse = await fetch(resourceMetadataUrl);
		if (!resourceResponse.ok) {
			console.warn(`[addon-mcp] Failed to fetch OAuth resource metadata from ${resourceMetadataUrl}: ${resourceResponse.status}`);
			return null;
		}
		const resourceResult = v.safeParse(OAuthResourceMetadata, await resourceResponse.json());
		if (!resourceResult.success) {
			console.warn(`[addon-mcp] Invalid OAuth resource metadata from ${resourceMetadataUrl}: ${resourceResult.issues.map((i) => i.message).join(", ")}`);
			return null;
		}
		const serverMetadataUrl = `${resourceResult.output.authorization_servers[0]}/.well-known/oauth-authorization-server`;
		const serverResponse = await fetch(serverMetadataUrl);
		if (!serverResponse.ok) {
			console.warn(`[addon-mcp] Failed to fetch OAuth server metadata from ${serverMetadataUrl}: ${serverResponse.status}`);
			return null;
		}
		const serverResult = v.safeParse(OAuthServerMetadata, await serverResponse.json());
		if (!serverResult.success) {
			console.warn(`[addon-mcp] Invalid OAuth server metadata from ${serverMetadataUrl}: ${serverResult.issues.map((i) => i.message).join(", ")}`);
			return null;
		}
		return {
			resourceMetadataUrl,
			resourceMetadata: resourceResult.output,
			serverMetadata: serverResult.output
		};
	}
};
/**
* Extract Bearer token from Authorization header.
* Handles both Node.js (string | string[] | undefined) and Web API (string | null) headers.
*/
function extractBearerToken(authHeader) {
	const bearer = (Array.isArray(authHeader) ? authHeader : [authHeader]).find((value) => typeof value === "string" && value.startsWith("Bearer "));
	return bearer ? bearer.slice(7) : null;
}

//#endregion
//#region src/auth/get-refs-from-config.ts
/**
* Get composed Storybook refs from Storybook config.
* See: https://storybook.js.org/docs/sharing/storybook-composition
*/
async function getRefsFromConfig(options) {
	try {
		const refs = await options.presets.apply("refs", {});
		if (!refs || typeof refs !== "object") return [];
		return Object.entries(refs).flatMap(([key, value]) => {
			if (!value || typeof value !== "object") return [];
			const { title, url } = value;
			if (typeof url !== "string" || url.length === 0) return [];
			return [{
				id: key,
				title: typeof title === "string" && title.length > 0 ? title : key,
				url
			}];
		});
	} catch {
		return [];
	}
}

//#endregion
//#region src/auth/resolve-composition-sources.ts
const MANIFEST_PROBE_TIMEOUT_MS = 3e3;
async function resolveCompositionSources(options) {
	const refs = await getRefsFromConfig(options);
	const compositionAuth = new CompositionAuth();
	if (refs.length === 0) return {
		refs,
		compositionAuth,
		sources: void 0,
		multiSource: false
	};
	await compositionAuth.initialize(refs);
	const sources = compositionAuth.buildSources();
	return {
		refs,
		compositionAuth,
		sources,
		multiSource: sources.some((source) => !!source.url)
	};
}
async function resolveServerlessCompositionSources(options) {
	const refs = await getRefsFromConfig(options);
	const sources = [{
		id: "local",
		title: "Local"
	}, ...(await Promise.all(refs.map(async (ref) => await hasServerlessComponentManifestSource(ref.url) ? ref : void 0))).filter((ref) => !!ref).map((ref) => ({
		id: ref.id,
		title: ref.title,
		url: ref.url
	}))];
	return {
		refs,
		sources,
		multiSource: sources.some((source) => !!source.url)
	};
}
async function hasServerlessComponentManifestSource(refUrl) {
	const controller = new AbortController();
	const timeout = setTimeout(() => controller.abort(), MANIFEST_PROBE_TIMEOUT_MS);
	try {
		const response = await fetch(new URL("manifests/components.json", `${refUrl.replace(/\/$/, "")}/`), {
			headers: { Accept: "application/json" },
			signal: controller.signal
		});
		if (response.status === 401 && hasOAuthResourceMetadataChallenge(response)) return true;
		if (!response.ok) return false;
		const text = await response.text();
		return v.safeParse(v.pipe(v.string(), v.parseJson(), ComponentManifestMap), text).success;
	} catch {
		return false;
	} finally {
		clearTimeout(timeout);
	}
}
function hasOAuthResourceMetadataChallenge(response) {
	return /resource_metadata="[^"]+"/.test(response.headers.get("WWW-Authenticate") ?? "");
}

//#endregion
//#region src/storybook-ai-metadata.ts
const createEmptyInputSchema = () => ({
	type: "object",
	properties: {}
});
async function buildStorybookAiMetadata(options, existingMetadata) {
	const toolsets = v.parse(AddonOptions, {
		endpoint: options.endpoint,
		toolsets: options.toolsets ?? {}
	}).toolsets;
	const features = await options.presets.apply("features", {});
	const devEnabled = toolsets?.dev ?? true;
	const rawAvailability = await getToolAvailability(options, {
		features,
		moduleGraphSupported: await isModuleGraphSupportedByBuilder(options)
	});
	const availability = {
		...rawAvailability,
		reviewEnabled: rawAvailability.reviewEnabledForCli
	};
	const testEnabled = (toolsets?.test ?? true) && availability.testSupported;
	const docsToolsetEnabled = toolsets?.docs ?? true;
	const multiSource = docsToolsetEnabled ? (await resolveServerlessCompositionSources(options)).multiSource : false;
	const registryAvailability = getEffectiveToolAvailability(availability, { multiSource });
	const docsEnabled = docsToolsetEnabled && registryAvailability.docsEnabled;
	const registryContext = {
		availability: registryAvailability,
		multiSource,
		toolsets
	};
	const toolMetadata = getAddonToolMetadata(registryContext);
	const localTools = {
		...existingMetadata?.localTools,
		...getAddonLocalTools({
			...registryContext,
			options
		})
	};
	return {
		instructions: joinInstructions(existingMetadata?.instructions, buildServerInstructions({
			devEnabled,
			testEnabled,
			docsEnabled,
			changeDetectionEnabled: availability.changeDetectionEnabled,
			moduleGraphSupported: availability.moduleGraphSupported,
			reviewEnabled: availability.reviewEnabled
		})),
		tools: mergeToolDescriptors(existingMetadata?.tools ?? [], await toToolDescriptors(toolMetadata)),
		localTools
	};
}
async function toToolDescriptors(toolMetadata) {
	const adapter = new ValibotJsonSchemaAdapter();
	return Promise.all(toolMetadata.map(async ({ schema, outputSchema, ...metadata }) => ({
		...metadata,
		inputSchema: schema ? await toJsonSchema(adapter, schema) : createEmptyInputSchema(),
		...outputSchema ? { outputSchema: await toJsonSchema(adapter, outputSchema) } : {}
	})));
}
async function toJsonSchema(adapter, schema) {
	return await adapter.toJsonSchema(schema);
}
function joinInstructions(...sections) {
	const uniqueSections = [];
	const seen = /* @__PURE__ */ new Set();
	for (const section of sections) {
		const trimmed = section?.trim();
		if (trimmed && !seen.has(trimmed)) {
			seen.add(trimmed);
			uniqueSections.push(trimmed);
		}
	}
	return uniqueSections.join("\n\n");
}
function mergeToolDescriptors(...groups) {
	const toolsByName = /* @__PURE__ */ new Map();
	for (const tool of groups.flat()) {
		if (toolsByName.has(tool.name)) toolsByName.delete(tool.name);
		toolsByName.set(tool.name, tool);
	}
	return [...toolsByName.values()];
}

//#endregion
//#region src/manifests/vendored.ts
/** System tags used by Storybook for categorizing stories and docs entries. */
const Tag = {
	ATTACHED_MDX: "attached-mdx",
	UNATTACHED_MDX: "unattached-mdx",
	MANIFEST: "manifest"
};
/** Open-service ids (also the on-disk directory names under `services/`). */
const DOCGEN_SERVICE_ID = "core/docgen";
const STORY_DOCS_SERVICE_ID = "core/story-docs";
const MDX_SERVICE_ID = "addon-docs/mdx";
/**
* Derives the componentId portion of a story index entry id. Story ids have the
* shape `<componentId>--<storyName>`; the prefix before the first `--` is the
* stable component identifier.
*/
function getComponentIdFromEntry(entry) {
	return entry.id.split("--")[0] ?? entry.id;
}
function isAttachedDocsEntry(entry) {
	return entry.type === "docs" && entry.tags?.includes(Tag.ATTACHED_MDX) === true && entry.storiesImports.length > 0;
}
function isEligibleStoryEntry(entry) {
	return entry.type === "story" && entry.subtype === "story";
}
/**
* Picks one index entry per componentId: story entries win; attached docs fill gaps
* only where no story exists for that componentId.
*/
function selectComponentEntriesByComponentId(indexEntries) {
	const entriesByComponentId = /* @__PURE__ */ new Map();
	for (const entry of indexEntries) {
		if (!isEligibleStoryEntry(entry)) continue;
		entriesByComponentId.set(getComponentIdFromEntry(entry), entry);
	}
	for (const entry of indexEntries) {
		if (!isAttachedDocsEntry(entry)) continue;
		const componentId = getComponentIdFromEntry(entry);
		if (!entriesByComponentId.has(componentId)) entriesByComponentId.set(componentId, entry);
	}
	return entriesByComponentId;
}
/** `$ref` target for one component's story-docs payload, relative to `manifests/`. */
function storyDocsManifestRef(id) {
	return `../services/${STORY_DOCS_SERVICE_ID}/${id}.json#/components/${id}`;
}

//#endregion
//#region src/manifests/service-access.ts
let cachedGetService;
/**
* Resolves `getService` once via a dynamic import. Returns `undefined` on Storybook
* versions that don't export it (so callers fall back to the fetch-based path).
*/
async function loadGetService() {
	if (cachedGetService !== void 0) return cachedGetService ?? void 0;
	try {
		const getService = (await import("storybook/internal/core-server")).getService;
		cachedGetService = getService ?? null;
		return getService;
	} catch {
		cachedGetService = null;
		return;
	}
}
async function getServiceSafely(id) {
	const getService = await loadGetService();
	if (!getService) return;
	try {
		return getService(id);
	} catch {
		return;
	}
}
function getDocgenService() {
	return getServiceSafely(DOCGEN_SERVICE_ID);
}
function getStoryDocsService() {
	return getServiceSafely(STORY_DOCS_SERVICE_ID);
}
function getMdxService() {
	return getServiceSafely(MDX_SERVICE_ID);
}

//#endregion
//#region src/manifests/in-process-provider.ts
function classifyIndex(index) {
	const entries = Object.values(index.entries).filter((entry) => entry.tags?.includes(Tag.MANIFEST) ?? false);
	const selected = selectComponentEntriesByComponentId(entries);
	const storyBasedIds = /* @__PURE__ */ new Set();
	for (const [id, entry] of selected) if (entry.type === "story") storyBasedIds.add(id);
	const attachedDocsByComponent = /* @__PURE__ */ new Map();
	const unattachedDocs = /* @__PURE__ */ new Map();
	for (const entry of entries) {
		if (entry.type !== "docs") continue;
		if (entry.tags?.includes(Tag.UNATTACHED_MDX)) unattachedDocs.set(entry.id, entry);
		else if (entry.tags?.includes(Tag.ATTACHED_MDX)) {
			const componentId = getComponentIdFromEntry(entry);
			const list = attachedDocsByComponent.get(componentId) ?? [];
			list.push(entry);
			attachedDocsByComponent.set(componentId, list);
		}
	}
	return {
		componentIds: [...selected.keys()],
		storyBasedIds,
		attachedDocsByComponent,
		unattachedDocs
	};
}
/**
* Builds the shallow `components.json` surface used by `list-all-documentation`.
*
* In dev, single component/doc lookups use `resolveEntry`, so this index does not
* carry docgen or MDX refs. The only ref it exposes is `stories`, which
* `list-all-documentation --withStoryIds` follows to avoid running full docgen.
*/
async function buildComponentsManifest(cls) {
	const docgenService = await getDocgenService();
	const allDocgen = docgenService ? await docgenService.queries.docgenForAllComponents.loaded() : {};
	const components = {};
	for (const id of cls.componentIds) {
		const payload = allDocgen[id];
		components[id] = {
			id,
			name: payload?.name ?? id,
			...payload?.description !== void 0 ? { description: payload.description } : {},
			...payload?.summary !== void 0 ? { summary: payload.summary } : {},
			...cls.storyBasedIds.has(id) ? { stories: { $ref: storyDocsManifestRef(id) } } : {}
		};
	}
	return JSON.stringify({
		v: 1,
		components
	});
}
/** Builds the ref-based unattached `docs.json` index, or `undefined` when there are none. */
async function buildDocsManifest(cls) {
	if (cls.unattachedDocs.size === 0) return;
	const mdxService = await getMdxService();
	const allMdx = mdxService ? await mdxService.queries.mdxForAllComponents.loaded() : {};
	const docs = {};
	for (const [docId, entry] of cls.unattachedDocs) {
		const payload = allMdx[docId]?.docs?.[docId];
		docs[docId] = {
			id: docId,
			name: entry.name,
			...payload?.summary !== void 0 ? { summary: payload.summary } : {}
		};
	}
	return JSON.stringify({
		v: 1,
		docs
	});
}
const STORY_DOCS_SERVICE_PREFIX = `services/${STORY_DOCS_SERVICE_ID}/`;
function matchStoryDocsServicePath(normalizedPath) {
	if (normalizedPath.startsWith(STORY_DOCS_SERVICE_PREFIX) && normalizedPath.endsWith(".json")) return normalizedPath.slice(STORY_DOCS_SERVICE_PREFIX.length, -5);
}
/** Wraps one story-docs payload as `{ components: { [id]: payload } }`. */
async function buildStoryDocsServiceFile(id) {
	const storyDocsService = await getStoryDocsService();
	const payload = storyDocsService ? await storyDocsService.queries.storyDocs.loaded({ id }) : void 0;
	return JSON.stringify({ components: payload === void 0 ? {} : { [id]: payload } });
}
async function resolveComponent(id, cls) {
	const [docgenService, storyDocsService, mdxService] = await Promise.all([
		getDocgenService(),
		getStoryDocsService(),
		getMdxService()
	]);
	const [docgen, storyDocs] = await Promise.all([docgenService ? docgenService.queries.docgen.loaded({ id }) : void 0, storyDocsService ? storyDocsService.queries.storyDocs.loaded({ id }) : void 0]);
	const attached = cls.attachedDocsByComponent.get(id) ?? [];
	let docs;
	if (attached.length > 0 && mdxService) {
		const mdxPayload = await mdxService.queries.mdxForComponent.loaded({ id });
		if (mdxPayload?.docs) {
			docs = {};
			for (const entry of attached) {
				const doc = mdxPayload.docs[entry.id];
				if (doc) docs[entry.id] = doc;
			}
		}
	}
	return {
		kind: "component",
		component: adaptCoreComponent({
			...docgen,
			id,
			name: docgen?.name ?? id,
			...storyDocs?.stories ? { stories: storyDocs.stories } : {},
			...storyDocs?.import ? { import: storyDocs.import } : {},
			...docs ? { docs } : {}
		})
	};
}
async function resolveStandaloneDoc(id) {
	const mdxService = await getMdxService();
	const doc = (mdxService ? await mdxService.queries.mdxForComponent.loaded({ id }) : void 0)?.docs?.[id];
	if (!doc) return;
	return {
		kind: "doc",
		doc: adaptCoreDoc(doc)
	};
}
/**
* Builds the in-process manifest provider and single-entry resolver for
* docgen-server mode. The closures read the live story index and services on each
* call (both internally cached), so they stay in lock-step with HMR.
*/
function createDocgenServerManifestAccess(options) {
	const manifestProvider = async (_request, path) => {
		const normalized = path.replace(/^\.?\//, "");
		if (normalized === "manifests/components.json") return buildComponentsManifest(classifyIndex(await getStoryIndex(options)));
		if (normalized === "manifests/docs.json") {
			const docs = await buildDocsManifest(classifyIndex(await getStoryIndex(options)));
			if (docs === void 0) throw new Error("No unattached docs manifest available");
			return docs;
		}
		const storyDocsId = matchStoryDocsServicePath(normalized);
		if (storyDocsId) return buildStoryDocsServiceFile(storyDocsId);
		throw new Error(`Unsupported in-process manifest path: ${path}`);
	};
	const resolveEntry = async (id) => {
		const cls = classifyIndex(await getStoryIndex(options));
		if (cls.unattachedDocs.has(id)) return resolveStandaloneDoc(id);
		if (cls.componentIds.includes(id)) return resolveComponent(id, cls);
	};
	return {
		manifestProvider,
		resolveEntry
	};
}

//#endregion
//#region src/preset.ts
const previewAnnotations = async (existingAnnotations = []) => {
	return [...existingAnnotations, path.join(import.meta.dirname, "preview.js")];
};
const experimental_devServer = async (app, options) => {
	const addonOptions = v.parse(AddonOptions, {
		endpoint: options.endpoint,
		toolsets: options.toolsets ?? {}
	});
	const origin = `http://localhost:${options.port}`;
	const endpoint = addonOptions.endpoint ?? DEFAULT_MCP_ENDPOINT;
	const { refs, compositionAuth, sources, multiSource } = await resolveCompositionSources(options);
	const rawAvailability = await getToolAvailability(options);
	const docgenServerAccess = rawAvailability.docgenServer ? createDocgenServerManifestAccess(options) : void 0;
	let createManifestProvider;
	if (refs.length > 0) {
		logger.info(`Initialized composition with ${refs.length} remote Storybook(s)`);
		if (compositionAuth.requiresAuth) logger.info(`Auth required for: ${compositionAuth.authUrls.join(", ")}`);
		logger.info(`Sources: ${(sources ?? []).map((s) => s.id).join(", ")}`);
		createManifestProvider = () => compositionAuth.createManifestProvider(origin, docgenServerAccess?.manifestProvider);
	}
	const manifestAccessFor = (req) => ({
		manifestProvider: createManifestProvider ? createManifestProvider(req) : docgenServerAccess?.manifestProvider,
		resolveEntry: docgenServerAccess?.resolveEntry
	});
	app.get("/.well-known/oauth-protected-resource", (_req, res) => {
		const wellKnown = compositionAuth.buildWellKnown(origin);
		if (!wellKnown) {
			res.writeHead(404);
			res.end("Not found");
			return;
		}
		res.writeHead(200, { "Content-Type": "application/json" });
		res.end(JSON.stringify(wellKnown));
	});
	const requireAuth = (req, res) => {
		const token = extractBearerToken(req.headers["authorization"]);
		if (compositionAuth.requiresAuth && !token) {
			res.writeHead(401, {
				"Content-Type": "text/plain",
				"WWW-Authenticate": compositionAuth.buildWwwAuthenticate(origin)
			});
			res.end("401 - Unauthorized");
			return true;
		}
		return false;
	};
	app.post(endpoint, (req, res) => {
		if (requireAuth(req, res)) return;
		return mcpServerHandler({
			req,
			res,
			options,
			addonOptions,
			endpoint,
			sources,
			...manifestAccessFor(req),
			compositionAuth
		});
	});
	const { moduleGraphSupported, changeDetectionEnabled, reviewEnabled, reviewEnabledForCli, docsEnabled, docsHasManifests, docsFeatureEnabled, testSupported, a11yEnabled } = getEffectiveToolAvailability(rawAvailability, { multiSource });
	const isDevEnabled = addonOptions.toolsets?.dev ?? true;
	const isDocsEnabled = docsEnabled && (addonOptions.toolsets?.docs ?? true);
	const isTestEnabled = testSupported && (addonOptions.toolsets?.test ?? true);
	app.get(endpoint, (req, res) => {
		if (!req.headers["accept"]?.includes("text/html")) {
			if (requireAuth(req, res)) return;
			return mcpServerHandler({
				req,
				res,
				options,
				addonOptions,
				endpoint,
				sources,
				...manifestAccessFor(req),
				compositionAuth
			});
		}
		res.writeHead(200, { "Content-Type": "text/html" });
		let docsNotice = "";
		if (!docsHasManifests) docsNotice = `<div class="toolset-notice">
				This toolset is only supported in React-based setups.
			</div>`;
		else if (!docsFeatureEnabled) docsNotice = `<div class="toolset-notice">
				This toolset requires enabling the component manifest feature.
				<a target="_blank" href="https://github.com/storybookjs/mcp/tree/main/packages/addon-mcp#docs-tools-experimental">Learn how to enable it</a>
			</div>`;
		const testNoticeLines = [!testSupported && `This toolset requires Storybook 10.3.0+ with <code>@storybook/addon-vitest</code>. <a target="_blank" href="https://storybook.js.org/docs/writing-tests/integrations/vitest-addon#install-and-set-up">Learn how to set it up</a>`, !a11yEnabled && `Add <code>@storybook/addon-a11y</code> for accessibility testing. <a target="_blank" href="https://storybook.js.org/docs/writing-tests/accessibility-testing">Learn more</a>`].filter(Boolean);
		const testNotice = testNoticeLines.length ? `<div class="toolset-notice">${testNoticeLines.join("<br>")}</div>` : "";
		const a11yBadge = a11yEnabled ? " <span class=\"toolset-status enabled\">+ accessibility</span>" : "";
		const devNoticeLines = !isDevEnabled ? [`The <code>dev</code> toolset is disabled via addon options.`] : [
			!moduleGraphSupported && `<code>get-stories-by-component</code> requires a dev server with a builder that supports the module graph (e.g. Vite).`,
			!changeDetectionEnabled && `<code>get-changed-stories</code> requires enabling the <code>changeDetection</code> feature flag.`,
			!reviewEnabled && (reviewEnabledForCli ? `<code>display-review</code> is enabled for <code>storybook ai</code> CLI clients (the Claude/Codex plugins); direct MCP clients need the <code>experimentalReview</code> feature flag.` : `<code>display-review</code> requires the <code>changeDetection</code> feature flag and is off when <code>experimentalReview</code> is set to <code>false</code>.`)
		].filter(Boolean);
		const devNotice = devNoticeLines.length ? `<div class="toolset-notice">${devNoticeLines.join("<br>")}</div>` : "";
		const statusWord = (enabled) => enabled ? "enabled" : "disabled";
		const html = template_default.replaceAll("{{DEV_STATUS}}", isDevEnabled ? "enabled" : "disabled").replaceAll("{{STORIES_BY_COMPONENT_STATUS}}", statusWord(isDevEnabled && moduleGraphSupported)).replaceAll("{{CHANGE_DETECTION_STATUS}}", statusWord(isDevEnabled && changeDetectionEnabled)).replaceAll("{{REVIEW_STATUS}}", statusWord(isDevEnabled && reviewEnabledForCli)).replace("{{DEV_NOTICE}}", devNotice).replaceAll("{{DOCS_STATUS}}", isDocsEnabled ? "enabled" : "disabled").replace("{{DOCS_NOTICE}}", docsNotice).replaceAll("{{TEST_STATUS}}", isTestEnabled ? "enabled" : "disabled").replace("{{TEST_NOTICE}}", testNotice).replace("{{MANIFEST_DEBUGGER_LINK}}", rawAvailability.docsEnabled ? "<p>View the <a href=\"/manifests/components.html\">component manifest debugger</a>.</p>" : "").replace("{{A11Y_BADGE}}", a11yBadge);
		res.end(html);
	});
	return app;
};
const experimental_storybookAi = async (existingMetadata, options) => {
	return buildStorybookAiMetadata(options, existingMetadata);
};
const features = async (existingFeatures) => {
	return {
		...existingFeatures,
		componentsManifest: true
	};
};

//#endregion
export { experimental_devServer, experimental_storybookAi, features, previewAnnotations };
